{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Thymus ageing | B cell compartment: Load IG data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys \n",
    "import os\n",
    "from datetime import datetime\n",
    "today = datetime.today().strftime('%Y-%m-%d')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import hdf5plugin\n",
    "import dandelion as ddl\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Add repo path to sys path (allows to access scripts and metadata from repo)\n",
    "repo_path = '/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/B_compartment'\n",
    "sys.path.insert(1, repo_path) \n",
    "sys.path.insert(2, '/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts')\n",
    "\n",
    "# Add R libs path\n",
    "os.environ['R_LIBS_USER'] = '/nfs/team205/lm25/condaEnvs/thymusAgeing/lib/R/library'\n",
    "\n",
    "# Formatting\n",
    "from matplotlib import font_manager\n",
    "font_manager.fontManager.addfont(\"/nfs/team205/ny1/ThymusSpatialAtlas/software/Arial.ttf\")\n",
    "plt.style.use('/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts/plotting/thyAgeing.mplstyle')\n",
    "\n",
    "%load_ext rpy2.ipython\n",
    "%reload_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture output\n",
    "%%R\n",
    "\n",
    "library(tidyverse)\n",
    "library(patchwork)\n",
    "library(magrittr)\n",
    "\n",
    "source('/nfs/team205/lm25/customScripts/visualisation/customTheme.R')\n",
    "\n",
    "options(max.print=150)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define plot and path dirs\n",
    "plots_path = os.path.join(repo_path, 'plots')\n",
    "data_path = os.path.join(repo_path, 'data')\n",
    "model_path = os.path.join(repo_path, 'models')\n",
    "general_data_path = '/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/data'\n",
    "\n",
    "print(f'Plot path: {plots_path}')\n",
    "print(f'Data path: {data_path}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotting.utils import plot_grouped_boxplot,calc_figsize,thyAgeing_colors,thyAgeing_greys\n",
    "from utils import get_latest_version,update_obs,freq_by_donor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in BCR data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update metadata\n",
    "from utils import get_latest_version,update_obs\n",
    "\n",
    "latest_meta_path = get_latest_version(dir = f'{general_data_path}/metadata', file_prefix='Thymus_ageing_metadata')\n",
    "meta = pd.read_excel(latest_meta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BCR\n",
    "bcr = []\n",
    "no_bcr = {'missing_dandelion' : [],\n",
    "          'missing_raw' : []}\n",
    "bcr_dir = '/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_general/thymus_vdj/BCR'\n",
    "for lib in meta['library_bcr'].tolist():\n",
    "    \n",
    "    if not pd.isna(lib):\n",
    "        bcr_path = f'{bcr_dir}/{lib}'  \n",
    "\n",
    "        if os.path.exists(bcr_path):\n",
    "            if 'dandelion' in os.listdir(bcr_path):\n",
    "                vdj = ddl.read_10x_airr(f'{bcr_path}/dandelion/all_contig_dandelion.tsv')\n",
    "                bcr.append(vdj)\n",
    "            else:\n",
    "                no_bcr['missing_dandelion'].append(lib)\n",
    "                print(f'No BCR dandelion results folder for {lib}')\n",
    "        else:\n",
    "            no_bcr['missing_raw'].append(lib)\n",
    "            print(f'No BCR path for {lib}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of BCR libraries\n",
    "len(bcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pprint\n",
    "\n",
    "pprint.pprint(no_bcr) # Missing raw is due to wrong library type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddl.concat(bcr).write(f'{data_path}/objects/vdj/thymusAgeing_bcr_v2_{today}.h5ddl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Aligning BCR data with B cell adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load data\n",
    "object_version = 'v4_2024-11-06'\n",
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}.zarr')\n",
    "vdj_version = 'v3_2025-02-19'\n",
    "bcr = ddl.read_h5ddl(f'{data_path}/objects/vdj/thymusAgeing_bcr_v2_{today}.h5ddl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check whether there are overlapping barcodes\n",
    "np.intersect1d(np.array(bcr.metadata.index), adata.obs_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BCR: Align bcr and adata\n",
    "bcr_filtered, adata_filtered = ddl.pp.check_contigs(bcr, adata, productive_only = False, library_type = 'ig')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save new metadata\n",
    "adata_filtered.obs.loc[:, 'has_contig':'rearrangement_status_VJ'].to_csv(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}_bcr_{vdj_version}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcr_filtered.write(f'{data_path}/objects/vdj/thymusAgeing_bcrFiltered_{vdj_version}.h5ddl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of BCRs per study\n",
    "adata_filtered.obs.loc[adata_filtered.obs.has_contig != 'No_contig'].groupby('study').size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BCR properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load VDJ data\n",
    "vdj_version = 'v3_2025-02-19'\n",
    "bcr = ddl.read_h5ddl(f'{data_path}/objects/vdj/thymusAgeing_bcrFiltered_{vdj_version}.h5ddl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata\n",
    "object_version = 'v4_2024-11-06'\n",
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}.zarr')\n",
    "\n",
    "# # Add BCR meta\n",
    "# bcr_meta = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}_bcr_{vdj_version}.csv', index_col = 0)\n",
    "# adata.obs = adata.obs.join(bcr_meta)\n",
    "\n",
    "# adata = adata[~pd.isna(adata.obs.has_contig), :]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clonotype calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Update metadata\n",
    "from utils import get_latest_version,update_obs\n",
    "\n",
    "latest_meta_path = get_latest_version(dir = f'{general_data_path}/metadata', file_prefix='Thymus_ageing_metadata')\n",
    "latest_meta = pd.read_excel(latest_meta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add donor info to BCR data\n",
    "bcr.data = bcr.data.merge(latest_meta[['donor', 'library_bcr']], left_on = 'sample_id', right_on = 'library_bcr')\n",
    "bcr.metadata = bcr.metadata.merge(latest_meta[['donor', 'library_bcr']], left_on = 'sample_id', right_on = 'library_bcr')\n",
    "bcr.data['donor'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check all seqs have donor info\n",
    "bcr.data['donor'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Automatically determine threshold\n",
    "ddl.pp.calculate_threshold(bcr,\n",
    "                           mode = 'single-cell',\n",
    "                           fields = 'donor',\n",
    "                           threshold_method = 'gmm',\n",
    "                           threshold_model = 'gamma-gamma',\n",
    "                           cutoff = 'optimal',\n",
    "                           model = 'hh_s1f', # human Ig-specific single nucleotide model similar to a transition/transversion model\n",
    "                           ncpu = 4, \n",
    "                           save_plot = f'{plots_path}/preprocessing/vdj/thyAgeing_bSplit_bcr_changeoThreshold.png',\n",
    "                           figsize = (20, 10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define clones using changeo\n",
    "ddl.tl.define_clones(bcr, \n",
    "                     model = 'hh_s1f',\n",
    "                     doublets = 'count',\n",
    "                     key_added=\"changeo_clone_id\",\n",
    "                     additional_args = ['--gf', 'donor'], \n",
    "                     ncpu = 4,\n",
    "                     out_dir = f'{data_path}/analysis/vdj/clonotypes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcr.metadata.groupby('changeo_clone_id').size().sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddl.tl.clone_size(bcr, clone_key = 'changeo_clone_id', key_added = 'changeo_clone_size')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddl.tl.generate_network(bcr, clone_key = 'changeo_clone_id', network_key = 'changeo_network', num_cores = 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mutation quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check germline mask is present in data\n",
    "bcr.data[[\"v_call_genotyped\", \"germline_alignment_d_mask\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Quantify mutational load\n",
    "ddl.pp.quantify_mutations(bcr, \n",
    "                          combine=False, # Returns silent and replacement mutations separately\n",
    "                          split_locus = True # Returns results for each locus separately\n",
    "                          )\n",
    "\n",
    "# Quantify mutational load\n",
    "ddl.pp.quantify_mutations(bcr, \n",
    "                          combine=True, # Returns silent and replacement mutations separately\n",
    "                          split_locus = True # Returns results for each locus separately\n",
    "                          )\n",
    "\n",
    "# Quantify mutational load\n",
    "ddl.pp.quantify_mutations(bcr, \n",
    "                          combine=True, # Returns silent and replacement mutations separately\n",
    "                          split_locus = True # Returns results for each locus separately\n",
    "                          )\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clonal diversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddl.pl.clone_rarefaction(adata, color=\"donor\", clone_key=\"changeo_clone_id\", figsize = (10, 5),\n",
    "                         save = f'{plots_path}/preprocessing/vdj/thyAgeing_bSplit_bcr_cloneRarefaction.png'\n",
    "                         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddl.tl.clone_diversity(\n",
    "    bcr, groupby=\"donor\", method=\"gini\", metric=\"clone_centrality\", clone_key=\"changeo_clone_id\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Transfer to adata and save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ddl.tl.transfer(adata, bcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save updated BCR data\n",
    "bcr.write(f'{data_path}/objects/vdj/thymusAgeing_bcrFiltered_{vdj_version}.h5ddl', compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save updated metadata\n",
    "adata.obs.loc[:, 'changeo_clone_id':'clone_centrality_gini'].to_csv(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}_bcr_{vdj_version}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[['isotype', 'isotype_status']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scanpy.plotting.palettes import default_28, default_102\n",
    "\n",
    "sc.set_figure_params(figsize=[4, 4])\n",
    "ddl.pl.clone_network(\n",
    "    adata,\n",
    "    color=[\n",
    "        \"donor\",\n",
    "        \"changeo_clone_size\",\n",
    "        \"mu_count_IGH\",\n",
    "        \"mu_count_IGL\",\n",
    "    ],\n",
    "    ncols=2,\n",
    "    legend_loc=\"none\",\n",
    "    legend_fontoutline=3,\n",
    "    edges_width=1,\n",
    "    palette=default_28 + default_102,\n",
    "    color_map=\"viridis\",\n",
    "    size=20,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.set_figure_params(figsize=[4, 4.5])\n",
    "ddl.pl.clone_network(\n",
    "    adata,\n",
    "    color=[\"locus_status\", \"chain_status\", 'isotype_status'],\n",
    "    ncols=2,\n",
    "    legend_fontoutline=3,\n",
    "    edges_width=1,\n",
    "    size=20,\n",
    "    wspace = 0.5,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Somatic hypermutation analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata\n",
    "object_version = 'v5_2025-04-16'\n",
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}.zarr')\n",
    "\n",
    "# Add new annotations to adata\n",
    "ct_anno = pd.read_csv(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_v4_2025-02-04_curatedAnno_v10.csv', index_col = 0)\n",
    "for c in [c for c in ct_anno.columns if c in adata.obs.columns]:\n",
    "    adata.obs.drop(columns = c, inplace = True)\n",
    "adata.obs = adata.obs.join(ct_anno, how = 'left')\n",
    "adata = adata[(adata.obs['anno_status'] == 'include') | (adata.obs['qc_status'] == 'PASS'), :]\n",
    "\n",
    "# Update metadata\n",
    "latest_meta_path = get_latest_version(dir = f'{general_data_path}/metadata', file_prefix='Thymus_ageing_metadata')\n",
    "latest_meta = pd.read_excel(latest_meta_path)\n",
    "update_obs(adata, latest_meta, on = 'index', ignore_warning = True)\n",
    "\n",
    "vdj_version = 'v3_2025-02-19'\n",
    "bcr = ddl.read_h5ddl(f'{data_path}/objects/vdj/thymusAgeing_bcrFiltered_{vdj_version}.h5ddl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify cell type columns\n",
    "from anno_levels import get_ct_levels,age_group_levels\n",
    "col_cell_type_broad = 'taa_l4'\n",
    "col_cell_type_fine = 'taa_l5'\n",
    "col_cell_type_broad_levels = get_ct_levels(level = col_cell_type_broad, taa_l1 = 'B')\n",
    "col_cell_type_fine_levels = get_ct_levels(level = col_cell_type_fine, taa_l1 = 'B')\n",
    "col_age_group = 'age_group'\n",
    "col_age_group_levels = eval(f'{col_age_group}_levels')\n",
    "\n",
    "%R -i plots_path,col_cell_type_broad,col_cell_type_fine,col_cell_type_broad_levels,col_cell_type_fine_levels,col_age_group,col_age_group_levels,data_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only include annotated and non-developing B cells\n",
    "adata = adata[~adata.obs['taa_l5'].isin(['B_GC-like-locnt','B_dev_thy','B_dev'])]\n",
    "bcr_subset = ddl.Dandelion(bcr.data[bcr.data['cell_id'].isin(adata.obs_names)], bcr.metadata)\n",
    "\n",
    "# Add donor column and write to airr\n",
    "bcr_subset.data = bcr_subset.data.merge(adata.obs[[col_age_group, col_cell_type_fine]], left_on = 'cell_id',right_index = True, how = 'left')\n",
    "bcr_subset.write_airr(filename=f'{data_path}/analyses/vdj/hypermutation/thymusAgeing_bcrFiltered_{vdj_version}_airr.tsv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform somatic hypermutation analysis using the immcantation framework, see:\n",
    "- https://immcantation.readthedocs.io/en/stable/getting_started/10x_tutorial.html\n",
    "- https://dowser.readthedocs.io/en/latest/vignettes/Resolve-Light-Chains-Vignette/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i data_path,vdj_version\n",
    "\n",
    "# Load bcr data\n",
    "bcr = read_tsv(file.path(data_path, paste0('/analyses/vdj/hypermutation/thymusAgeing_bcrFiltered_', vdj_version, '_airr.tsv')))\n",
    "\n",
    "# Remove any cells without a cell type and with stop codons\n",
    "bcr = bcr %>% filter(!is.na(!!rlang::sym(col_cell_type_fine)),\n",
    "                      !stop_codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "bcr %>% colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Remove cells with multiple heavy chains\n",
    "multi_heavy <- table(dplyr::filter(bcr, locus == \"IGH\")$cell_id)\n",
    "multi_heavy_cells <- names(multi_heavy)[multi_heavy > 1]\n",
    "\n",
    "print(paste('Removing', length(multi_heavy_cells), 'cells with multiple heavy chains'))\n",
    "\n",
    "bcr = bcr %>%\n",
    "dplyr::filter(!cell_id %in% multi_heavy_cells)\n",
    "\n",
    "nrow(bcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Remove cells without any heavy chain\n",
    "heavy_cells <- dplyr::filter(bcr, locus == \"IGH\")$cell_id\n",
    "light_cells <- dplyr::filter(bcr, locus == \"IGK\" | locus == \"IGL\")$cell_id\n",
    "no_heavy_cells <- light_cells[which(!light_cells %in% heavy_cells)]\n",
    "\n",
    "print(paste('Removing', length(no_heavy_cells), 'cells without heavy chain'))\n",
    "\n",
    "bcr = bcr %>%\n",
    "dplyr::filter(!cell_id %in% no_heavy_cells)\n",
    "\n",
    "nrow(bcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "bcr %>%\n",
    "rstatix::freq_table(donor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# calculate and plot the rank-abundance curve\n",
    "abund <- alakazam::estimateAbundance(bcr, clone = 'changeo_clone_id',group = \"donor\")\n",
    "\n",
    "abund_plot <- alakazam::plot(abund, silent=T)\n",
    "abund_plot\n",
    "ggsave(file.path(plots_path, 'vdjAnalysis/clonotypes/thyAgeing_bSplit_bcr_rankAbundance.png'), width = 10, height = 5, dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# # calculate and plot the rank-abundance curve\n",
    "# div <- alakazam::alphaDiversity(data.frame(abund@abundance), clone = 'changeo_clone_id',group = \"donor\", nboot = 100)\n",
    "# plot(div, silent=T) + facet_wrap(\"donor\", ncol = 3)\n",
    "# ggsave(file.path(plots_path, 'vdjAnalysis/clonotypes/thyAgeing_bSplit_bcr_diversity.png'), width = 10, height = 5, dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Remove sequences assigned to no clones\n",
    "bcr <- dplyr::filter(bcr, !is.na(changeo_clone_id))\n",
    "\n",
    "bcr <- dowser::resolveLightChains(bcr,\n",
    "                                  nproc = 4,\n",
    "                                  clone = \"changeo_clone_id\",\n",
    "                                  cell = \"cell_id\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# Remove sequences assigned to no clones\n",
    "bcr <- dplyr::filter(bcr, !is.na(changeo_clone_id))\n",
    "\n",
    "# Read in IMGT reference sequences\n",
    "references <- dowser::readIMGT(dir = \"/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/B_compartment/data/references/immcantation_imgt/human/vdj\")\n",
    "\n",
    "# Reconstruct germlines\n",
    "results <- dowser::createGermlines(bcr, references, \n",
    "                                   clone = 'clone_subgroup_id', # clone_subgroup_id for heavy and light paired analysis, changeo_clone_id for heavy only\n",
    "                                   fields = c(\"donor\"), \n",
    "                                   nproc = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Format clones\n",
    "clones <- dowser::formatClones(results,\n",
    "                               clone = 'changeo_clone_id',\n",
    "                               traits = c('donor'),\n",
    "                               text_fields = c(col_cell_type_fine),\n",
    "                               cell = 'cell_id',\n",
    "                               chain=\"HL\", # HL for paired analysis\n",
    "                               nproc=1, \n",
    "                               collapse = FALSE, \n",
    "                               split_light = TRUE, \n",
    "                               minseq = 3,\n",
    "                               )\n",
    "\n",
    "head(clones)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "\n",
    "# Save clones\n",
    "clones %>% saveRDS(file.path(data_path, paste0('analyses/vdj/hypermutation/thymusAgeing_bcrFiltered_', vdj_version,'_shmClones.rds')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "Sys.setenv(IGPHYML_PATH = \"/nfs/team205/lm25/condaEnvs/thymusAgeing/share/igphyml/motifs\")\n",
    "\n",
    "trees <- dowser::getTrees(clones, nproc = 8, build=\"igphyml\", exec=\"/nfs/team205/lm25/condaEnvs/thymusAgeing/bin/igphyml\", collapse = TRUE)\n",
    "\n",
    "trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "trees "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Add donor info and save\n",
    "trees <- trees %>%\n",
    "dplyr::left_join(bcr %>% dplyr::distinct(donor, clone_subgroup_id), by = c('clone_id' = 'clone_subgroup_id')) \n",
    "\n",
    "# Save clones\n",
    "trees %>% saveRDS(file.path(data_path, paste0('analyses/vdj/hypermutation/thymusAgeing_bcrFiltered_', vdj_version,'_shmTrees.rds')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "tree_plots <- dowser::plotTrees(trees, tips = \"taa_l5\", tipsize = 2)\n",
    "\n",
    "dowser::treesToPDF(tree_plots, file.path(plots_path, 'vdjAnalysis', 'clonotypes', 'thyAgeing_bSplit_bLineageTrees.pdf'), nrow = 2, ncol = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o trees_df\n",
    "\n",
    "trees_df <- trees %>% dplyr::select(clone_id, donor, seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trees_df = trees_df.merge(adata.obs[['donor', col_age_group, 'age_months']].drop_duplicates(), on = 'donor')\n",
    "\n",
    "trees_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trees_df.to_csv(f'{data_path}/analyses/vdj/hypermutation/thymusAgeing_bcrFiltered_{vdj_version}_shmTrees.csv', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = trees_df.groupby(['donor', 'age_group'], observed=True).size().reset_index(name='count')\n",
    "df_order = trees_df.drop_duplicates(subset = ['donor', 'age_months']).sort_values('age_months')['donor']\n",
    "df[col_age_group] = pd.Categorical(df[col_age_group], categories=['infant', 'paed', 'adult'], ordered=True)\n",
    "df[col_age_group].cat.remove_unused_categories()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotting.utils import calc_figsize,get_tint_palette,thyAgeing_colors\n",
    "from anno_levels import age_group_palette\n",
    "plt.figure(figsize=calc_figsize(width = 50, height = 35))\n",
    "sns.barplot(data = df, x = 'donor', y = 'count', hue = 'age_group', order = df_order, palette = get_tint_palette(thyAgeing_colors['magenta']),\n",
    "            linewidth = 0)\n",
    "plt.legend(title='Age group', bbox_to_anchor=(1.2, 1), loc='upper center')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Donor')\n",
    "plt.ylabel('N(SHM trees)')\n",
    "sns.despine(offset=2, trim = True)\n",
    "plt.yticks(ticks=np.arange(0, df['count'].max(), 5), labels=np.arange(0, df['count'].max(), 5))\n",
    "plt.savefig(f'{plots_path}/vdjAnalysis/clonotypes/thyAgeing_bSplit_bLineageTrees_count.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = trees_df.groupby(['donor', 'age_group'], observed=True).agg({'seqs' : 'sum'}).reset_index()\n",
    "df_order = trees_df.drop_duplicates(subset = ['donor', 'age_months']).sort_values('age_months')['donor']\n",
    "df[col_age_group] = pd.Categorical(df[col_age_group], categories=['infant', 'paed', 'adult'], ordered=True)\n",
    "df[col_age_group].cat.remove_unused_categories()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotting.utils import calc_figsize\n",
    "from anno_levels import age_group_palette\n",
    "from plotting.utils import calc_figsize,get_tint_palette,thyAgeing_colors\n",
    "from anno_levels import age_group_palette\n",
    "plt.figure(figsize=calc_figsize(width = 50, height = 35))\n",
    "sns.barplot(data = df, x = 'donor', y = 'seqs', hue = 'age_group', order = df_order, palette = get_tint_palette(thyAgeing_colors['magenta']),\n",
    "            linewidth = 0)\n",
    "plt.legend(title='Age group', bbox_to_anchor=(1.2, 1), loc='upper center')\n",
    "plt.xticks(rotation=45, ha='right')\n",
    "plt.xlabel('Donor')\n",
    "plt.ylabel('N(SHM trees)')\n",
    "sns.despine(offset=2, trim = True)\n",
    "plt.yticks(ticks=np.arange(0, df['seqs'].max(), 10), labels=np.arange(0, df['seqs'].max(), 10))\n",
    "plt.savefig(f'{plots_path}/vdjAnalysis/clonotypes/thyAgeing_bSplit_bLineageTreesCells_count.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# calculate switches along trees compared to 100 random permutations \n",
    "# this may take a while, and can be parallelized using nproc\n",
    "switches = dowser::findSwitches(trees, permutations=100, trait='donor', \n",
    "  igphyml=\"/nfs/team205/lm25/condaEnvs/thymusAgeing/bin/igphyml\", fixtrees=TRUE)\n",
    "\n",
    "ps = dowser::testPS(switches$switches)\n",
    "\n",
    "print(ps$means)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
