{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Thymus ageing atlas: B cell clonality and lineage analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys \n",
    "import os\n",
    "from datetime import datetime\n",
    "today = datetime.today().strftime('%Y-%m-%d')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import hdf5plugin\n",
    "import dandelion as ddl\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Add repo path to sys path (allows to access scripts and metadata from repo)\n",
    "repo_path = '/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/B_compartment'\n",
    "sys.path.insert(1, repo_path) \n",
    "sys.path.insert(2, '/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts')\n",
    "\n",
    "# Add R libs path\n",
    "os.environ['R_LIBS_USER'] = '/nfs/team205/lm25/condaEnvs/thymusAgeing/lib/R/library'\n",
    "\n",
    "%load_ext rpy2.ipython\n",
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# Define paths\n",
    "plots_path = f'{repo_path}/plots/'\n",
    "data_path = f'{repo_path}/data/'\n",
    "model_path = os.path.join(repo_path, 'models')\n",
    "general_data_path = '/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/data'\n",
    "\n",
    "print('Dir for plots: {}'.format(plots_path))\n",
    "print('Dir for data: {}'.format(data_path))\n",
    "\n",
    "# Formatting\n",
    "from matplotlib import font_manager\n",
    "font_manager.fontManager.addfont(\"/nfs/team205/ny1/ThymusSpatialAtlas/software/Arial.ttf\")\n",
    "plt.style.use('/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts/plotting/thyAgeing.mplstyle')\n",
    "\n",
    "# Import custom scripts\n",
    "from utils import get_latest_version,update_obs,freq_by_donor\n",
    "from anno_levels import get_ct_levels, get_ct_palette, age_group_levels, age_group_palette\n",
    "from plotting.utils import plot_grouped_boxplot, calc_figsize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture output\n",
    "%%R\n",
    "\n",
    "library(tidyverse)\n",
    "library(patchwork)\n",
    "library(magrittr)\n",
    "\n",
    "source('/nfs/team205/lm25/customScripts/visualisation/customTheme.R')\n",
    "\n",
    "options(max.print=150)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load and prepare data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata\n",
    "object_version = 'v4_2024-11-06'\n",
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}.zarr')\n",
    "\n",
    "# Add cell type annotation\n",
    "ct_anno = pd.read_csv(f'{data_path}/preprocessing/ctAnnotation/thyAgeing_bSplitxTissue_scvi_v2_2025-02-20_v5.csv', index_col=0)\n",
    "cols_overlapping = [col for col in ct_anno.columns if col in adata.obs.columns]\n",
    "if any(cols_overlapping):\n",
    "    adata.obs.drop(columns=cols_overlapping, inplace=True)\n",
    "adata.obs = adata.obs.join(ct_anno)\n",
    "\n",
    "# Update metadata\n",
    "latest_meta_path = get_latest_version(dir = f'{general_data_path}/metadata', file_prefix='Thymus_ageing_metadata')\n",
    "latest_meta = pd.read_excel(latest_meta_path)\n",
    "update_obs(adata, latest_meta, on = 'index', ignore_warning = True)\n",
    "\n",
    "vdj_version = 'v3_2025-02-19'\n",
    "bcr = ddl.read_h5ddl(f'{data_path}/objects/vdj/thymusAgeing_bcrFiltered_{vdj_version}.h5ddl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove explore level cells\n",
    "adata = adata[~adata.obs['taa_l5'].isna()]\n",
    "#adata = adata[~adata.obs['taa_l5'].str.contains('explore')]\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify cell type columns\n",
    "from anno_levels import get_ct_levels, age_group_levels\n",
    "col_cell_type_broad = 'taa_l3'\n",
    "col_cell_type_fine = 'taa_l5'\n",
    "col_cell_type_broad_levels = get_ct_levels(level = col_cell_type_broad, taa_l1 = 'B')\n",
    "col_cell_type_fine_levels = get_ct_levels(level = col_cell_type_fine, taa_l1 = 'B')\n",
    "col_age_group = 'age_group'\n",
    "col_age_group_levels = eval(f'{col_age_group}_levels')\n",
    "\n",
    "%R -i plots_path,col_cell_type_broad,col_cell_type_fine,col_cell_type_broad_levels,col_cell_type_fine_levels,col_age_group,col_age_group_levels,data_path"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explore BCR mutation level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcr_data = bcr.data.merge(adata.obs[[col_age_group, col_cell_type_fine]], left_on = 'cell_id', right_index = True, how = 'left')\n",
    "\n",
    "bcr_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = bcr_data[['mu_count', 'mu_freq', col_cell_type_fine, 'donor', col_age_group]].copy()\n",
    "df = df.loc[~df[col_cell_type_fine].str.contains('dev', na=True)]\n",
    "df[col_age_group] = pd.Categorical(df[col_age_group], categories = ['infant', 'paed', 'adult'], ordered = True)\n",
    "\n",
    "plt.figure(figsize = (calc_figsize(width = 70, height = 40)))\n",
    "sns.violinplot(data = df, x = col_cell_type_fine, y = 'mu_count', hue = col_age_group, inner = 'quartile', density_norm='width',\n",
    "               order = [c for c in col_cell_type_fine_levels if c in df[col_cell_type_fine].tolist()], palette = age_group_palette, scale = 'width')\n",
    "plt.xticks(rotation = 90)\n",
    "plt.xlabel('Cell type')\n",
    "plt.ylabel('N(mutations)')\n",
    "plt.legend(title = 'Age group', bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.savefig(f'{plots_path}/vdjAnalysis/clonotypes/thyAgeing_bSplit_bcrMutationCount_violin.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population markers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, groupby = 'taa_l4', n_genes = 100, use_raw = False, method = 't-test', key_added = f'rank_genes_taa_l4', reference = 'rest', pts = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes = {k:sc.get.rank_genes_groups_df(adata, group = k, key = f'rank_genes_taa_l4') for k in adata.obs['taa_l4'].unique().tolist() if not k == 'B_GC-like_explore'}\n",
    "marker_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes['B_GC-like'].head(50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- CCL22/CCL17 expressed in EBV infection and germinal centers (https://pubmed.ncbi.nlm.nih.gov/38574017/,https://pubmed.ncbi.nlm.nih.gov/33597749/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes['B_med'].head(30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- MS4A1: B cell proliferation/plasma cell differentiation\n",
    "- ARPC1B, ARPC2, ARPC3: role in B cell proliferation and controls BCR singalling (deficiency leads to autoimmune disorder), potential role in GCs -> antigen extraction ([Roper,2019](https://elifesciences.org/articles/48093))\n",
    "- CD74: B cell activation \n",
    "- RAC2: regulates cell adhesion in response to BCR stimulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "marker_genes['B_age-associated'].head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Extrafollicular vs GC AID+ cells: Extrafollicular AID+ cells were E47+ and Id2–, and occasionally cMYC+, JunB+, and CCL22+, and largely negative for IRF4 (Figure 5A-E). In the GC, AID+ cells were E47+ and Id2– (or borderline Id2+), and negative for cMYC, CCL22, IRF4, and JunB19,39,51 (and data not shown). Extrafollicular CD30+ cells uniquely shared some but not all of the AID+ cell phenotype, being virtually 100% cMYC+, JunB+, and Pax5+, and partially CD20+ and CCL22+ (in decreasing order of frequency). CD30+ cells were, however, IRF4+, Id2+, occasionally IgD+, and E47–. A transitional phenotype toward GC centroblasts was found in CD30+ GC cells (data not shown). (https://www.sciencedirect.com/science/article/pii/S0006497120638253)\n",
    "\n",
    "AIRE functions as checkpoint control in GC reaction, induced upon CD40 stimulation -> AIRE interacts with AIDA to reduce SHM (AIRE deficient mice have hypermutated antibodies) (https://pmc.ncbi.nlm.nih.gov/articles/PMC10802573/)\n",
    "\n",
    "CCL22/CCL17 is induced upon CD40 stimulation in GCs and promotes affinity maturation (https://www.nature.com/articles/s41586-021-03239-2)\n",
    "\n",
    "Extrathymic AIRE expression is regulated through IRF4, IRF8, TBX21, TCF7 and depends on NFkB signalling (plus RANK/CD40/LTB in mTECs) (https://www.sciencedirect.com/science/article/pii/S1568997222001112)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_markers = {'General': ['AICDA', 'AIRE', 'TNFRSF8', 'JUNB', 'MYC', 'TCF3', 'ID2', 'ID3', 'CD83', 'CD40', 'IRF8', 'TCF7'],\n",
    "              'GC': ['PAX5', 'BCL6', 'IRF4'],\n",
    "              'DZ' : ['CXCR4'],\n",
    "              #'LZ': ['CXCR5'],\n",
    "              #'Extrafollicular': ['TCF3'],\n",
    "              'EBV' : ['CCL22', 'CCL17', 'EBI3', 'CCL3', 'ICAM1'],\n",
    "              'Immunoglobulin': ['IGHM', 'IGHD', 'IGHA1', 'IGHA2','IGHG1', 'IGHG3', 'IGHE'],\n",
    "              'PC': ['PRDM1', 'XBP1', 'IRF4'],\n",
    "              'ABC': ['TBX21', 'RGS1', 'IFI44L', 'ALOX5AP']}\n",
    "\n",
    "b_markers = {'Other': ['IL6R', 'TNFRSF18','TNFSF13B'],\n",
    "             'DZ' : ['AICDA', 'CXCR4', 'MYC', 'MKI67', 'TOP2A', 'PCNA', 'BACH2', 'TCF3', 'PAX5', 'IRF4', 'MEF2B', 'FOXO1'],\n",
    "               'LZ' : ['CXCR5', 'CD83', 'CD86', 'MYBL1', 'SOCS3', 'CD40'],\n",
    "               'T cell contact': ['CXCL10', 'CCL5', 'CCL3'],\n",
    "               'GC_misc' : ['ID2', 'ID3',],\n",
    "               'BCR activtation' : ['CCL22', 'CCL17', 'EBI3', 'CCL3', 'ICAM1'],\n",
    "               'Recruitment' : ['CCR7'],\n",
    "               'B_naive' : ['CD22', 'SELL', 'IL4R', 'TCL1A', 'CR2', 'FOXO1', 'IGHM', 'IGHD',],\n",
    "               'B_mem' : ['CD27', 'CD38', 'FCRL4', 'FCRL5', 'CD44', 'PRDM1', 'IGHA1', 'IGHG1', 'IGHE'],\n",
    "               'B_plasma' : ['PRDM1', 'XBP1', 'MZB1'], \n",
    "               'B_age-assoc' : ['TBX21', 'ITGAX', 'IRF4'],\n",
    "               'B_pan' : ['CD19', 'MS4A1'],\n",
    "               'B_med' : ['HLA-DRA', 'HLA-DRB1','AIRE', 'IL15', 'LTA', 'LTB', 'PTPRC', 'CD5', 'SPN', 'CD80' ,'LY6G6C'],\n",
    "               'B_dev' : ['IGLL1', 'MME', 'RAG1'],\n",
    "               'B_dev_thy' : ['CD34', 'VPREB1', 'TYROBP'],\n",
    "               'B_mem_CD1C+' : ['CD1C', 'IGHM', 'FCRL3', 'CR2']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata, \n",
    "                b_markers,\n",
    "                categories_order=[c for c in col_cell_type_fine_levels if c in adata.obs[col_cell_type_fine].tolist()],\n",
    "                groupby = col_cell_type_fine,\n",
    "                figsize = calc_figsize(width = 250, height = 50),\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().style(smallest_dot=1, largest_dot = 40).savefig(f'{plots_path}/ctAnnotation/v6/thyAgeing_bSplit_bMarkersExt_dotplot.png')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Marker expression\n",
    "b_markers = {'DZ' : ['CXCR4', 'MYC', 'MKI67', 'TOP2A', 'AICDA', 'PCNA', 'BACH2'],\n",
    "               'LZ' : ['CXCR5', 'CD83', 'IRF4', 'CD86', 'MYBL1', 'SOCS3'],\n",
    "               'recruitment' : ['CCR7'],\n",
    "               'B_naive' : ['CD22', 'SELL', 'IL4R', 'TCL1A', 'CR2', 'FOXO1', 'IGHM', 'IGHD',],\n",
    "               'B_mem' : ['CD27', 'CD38', 'FCRL4', 'FCRL5', 'CD44', 'PRDM1', 'IGHA1', 'IGHG1', 'IGHE'],\n",
    "               'B_plasma' : ['PRDM1', 'XBP1', 'MZB1'], \n",
    "               'B_age-assoc' : ['TBX21', 'ITGAX',], # ITGAX = CD11c\n",
    "               'B_pan' : ['CD19', 'MS4A1'],\n",
    "               'B_med' : ['HLA-DRA', 'AIRE', 'IL15', 'LTA', 'LTB'],\n",
    "               'B_dev' : ['IGLL1', 'MME', 'RAG1', 'PAX5', 'EBF1', 'BCL11B'],\n",
    "               'B_dev_thy' : ['CD34', 'VPREB1', 'TYROBP',],}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata, \n",
    "                b_markers,\n",
    "                categories_order=[c for c in col_cell_type_fine_levels if c in adata.obs[col_cell_type_fine].tolist()],\n",
    "                groupby = col_cell_type_fine,\n",
    "                figsize = calc_figsize(width = 200, height = 50),\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().style(smallest_dot=1, largest_dot = 40).savefig(f'{plots_path}/ctAnnotation/v6/thyAgeing_bSplit_bMarkers_dotplot.pdf')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[c for c in]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Different stages of GC-like B cells: GC-like, GC-like_prolif, GC-like_activated (prev B_med, potentially related to EBV infection) -> Does EBV infection play a role? There are reports of GC-like fates of EBV infected cells (checking with Alex P whether we can investigate EBV transcript expression)\n",
    "\n",
    "CD40 stimulation induces AIRE: https://www.sciencedirect.com/science/article/pii/S2589004222021253?via%3Dihub, https://www.pnas.org/doi/10.1073/pnas.2220120120, https://www.sciencedirect.com/science/article/pii/S1074761315002113\n",
    "\n",
    "CD5 is marker of thymic B cells (https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2021.766698/full#B25)\n",
    "\n",
    "Paper suggesting that recirculating B cells can be licensced: https://www.sciencedirect.com/science/article/pii/S1074761315002113\n",
    "\n",
    "-> unsure of whether AIDA/AIRE B cells are resident thymic B cells or recirculating"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Are AIDA/AIRE expressing cells one population? -> likely yes, they co-cluster, share CD5 expression and have GC-like phenotype\n",
    "- Are these cells antigen presenting medullary B cells or recirculating B cells in GC reaction? Not sure currently. Medullary B cells are considered to present antigens to maturing thymocytes, partly driven through AIRE which is induced upon CD40 stimulation (https://www.sciencedirect.com/science/article/pii/S2589004222021253?via%3Dihub, https://www.pnas.org/doi/10.1073/pnas.2220120120, https://www.sciencedirect.com/science/article/pii/S1074761315002113). It is questionable whether these cells are of thymic origin only or whether recirculating B cells can also be licensced to present antigens (https://www.sciencedirect.com/science/article/pii/S1074761315002113). Medullary B cells also express CCL22 and CCL17, and undergo class switch recombination which requires AIDA expression, and have been reported to have a GC-like phenotype (https://www.sciencedirect.com/science/article/pii/S2589004222021253?via%3Dihub, https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2021.766698/full#B25). Age-associated loss of AIRE expression has also been reported (https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2021.766698/full#B25). On the other hand, GC B cells have been shown to express AIRE to dampen SHM upon strong BCR and CD40 stimulation (https://pmc.ncbi.nlm.nih.gov/articles/PMC10802573/). CCL22 and CCL17 expression have been described as a byproduct of strong BCR stimulation in GCs (https://www.nature.com/articles/s41586-021-03239-2). Furthermore, EBV infection has been reported to induce GC-like phenotypes and expression of CCL22, CCL17 and EBI3 (https://pubmed.ncbi.nlm.nih.gov/14747532/, https://pmc.ncbi.nlm.nih.gov/articles/PMC9554744/). -> AIDA expression causes SHM/CSR which happens both in GCs and medullary B cells // AIRE expression is induced upon CD40 stimulation and BCR signalling, triggered by thymocytes and T cells // CCL22 and CCL17 are signs of strong BCR activation\n",
    "- If this is one population arising from medullary B cells, is there a phenotype switch with age (more towards GC-like)? Need to investigate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_markers_filtered = {k: [v for v in gc_markers[k] if v in adata_spatial.var_names] for k in gc_markers.keys()}\n",
    "sc.pl.DotPlot(adata_spatial, \n",
    "                gc_markers_filtered,\n",
    "                groupby = 'taa_l4',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "moi = ['CXCR4', 'CXCR2', 'CXCR1', 'CXCR6', 'CXCL8', 'CXCL6', 'CXCL1',\n",
    "       'CXCL5', 'CXCL3', 'CXCL2', 'CXCL9', 'CXCL10', 'CXCL11', 'CXCL13',\n",
    "       'CXCL14', 'CXCL12', 'CXCR5', 'CXCL16', 'CXCL17', 'CXCR3',\n",
    "       'CCR4', 'CCR8', 'CCR9', 'CCR3', 'CCR1', 'CCR2', 'CCR5AS', 'CCR5',\n",
    "       'CCRL2', 'CCR6', 'CCR7', 'CCR10',\n",
    "       'CCL20', 'CCL28', 'CCL26', 'CCL24', 'CCL27', 'CCL19', 'CCL21',\n",
    "       'CCL22', 'CCL17', 'CCL2', 'CCL7', 'CCL11', 'CCL8', 'CCL13', 'CCL1',\n",
    "       'CCL5', 'CCL16', 'CCL14', 'CCL15', 'CCL23', 'CCL18', 'CCL3',\n",
    "       'CCL4', 'CCL3L1', 'CCL4L2', 'CCL25',\n",
    "       'IL22RA1', 'IL23R', 'IL12RB2', 'ILF2', 'IL6R-AS1', 'IL6R', 'ILDR2',\n",
    "       'IL10', 'IL19', 'IL20', 'IL24', 'IL1R2', 'IL1R1', 'IL1R1-AS1',\n",
    "       'IL1RL2', 'IL1RL1', 'IL18R1', 'IL18RAP', 'IL1A', 'IL1B', 'IL37',\n",
    "       'IL36G', 'IL36A', 'IL36B', 'IL36RN', 'IL1F10', 'IL1RN', 'ILKAP',\n",
    "       'IL5RA', 'IL17RE', 'IL17RC', 'IL17RB', 'IL17RD', 'ILDR1', 'IL20RB',\n",
    "       'IL20RB-AS1', 'IL12A-AS1', 'IL12A', 'IL1RAP', 'IL2', 'IL21',\n",
    "       'IL21-AS1', 'IL15', 'IL7R', 'IL31RA', 'IL6ST', 'IL3', 'IL5',\n",
    "       'IL13', 'IL4', 'IL9', 'IL17B', 'IL12B', 'ILRUN', 'IL17A', 'IL17F',\n",
    "       'IL20RA', 'IL22RA2', 'IL6', 'IL7', 'IL33', 'IL11RA', 'IL15RA',\n",
    "       'IL2RA', 'ILK', 'IL18BP', 'IL18', 'IL10RA', 'IL23A', 'IL26',\n",
    "       'IL22', 'IL31', 'IL17D', 'IL25', 'IL16', 'IL32', 'IL4R', 'IL21R',\n",
    "       'IL21R-AS1', 'IL27', 'IL34', 'IL17C', 'ILF3-DT', 'ILF3', 'IL27RA',\n",
    "       'ILVBL', 'IL12RB1', 'IL4I1', 'IL11', 'IL10RB-DT', 'IL10RB',\n",
    "       'IL17RA', 'IL2RB', 'IL17REL', 'IL3RA', 'IL1RAPL1', 'IL2RG',\n",
    "       'IL1RAPL2', 'IL13RA2', 'IL13RA1', 'IL9R',\n",
    "       'IFNLR1', 'IFNGR1', 'IFNB1', 'IFNW1', 'IFNA21', 'IFNA4', 'IFNA7',\n",
    "       'IFNA10', 'IFNA16', 'IFNA14', 'IFNA5', 'IFNA6', 'IFNA13', 'IFNA2',\n",
    "       'IFNA8', 'IFNA1', 'IFNE', 'IFNK', 'IFNG-AS1', 'IFNG', 'IFNL3',\n",
    "       'IFNL2', 'IFNL1', 'IFNAR2', 'IFNAR1', 'IFNGR2',\n",
    "       'TNFRSF18', 'TNFRSF4', 'TNFRSF14-AS1', 'TNFRSF14', 'TNFRSF25',\n",
    "       'TNFRSF9', 'TNFRSF8', 'TNFRSF1B', 'TNFAIP8L2', 'TNFSF18', 'TNFSF4',\n",
    "       'TNFAIP6', 'TNFSF10', 'TNFAIP8', 'TNF', 'TNFRSF21', 'TNFAIP3',\n",
    "       'TNFRSF10B', 'TNFRSF10C', 'TNFRSF10D', 'TNFRSF10A-AS1',\n",
    "       'TNFRSF10A', 'TNFRSF11B', 'TNFSF15', 'TNFSF8', 'TNFRSF1A',\n",
    "       'TNFRSF19', 'TNFSF11', 'TNFSF13B', 'TNFAIP2', 'TNFAIP8L3',\n",
    "       'TNFRSF12A', 'TNFRSF17', 'TNFSF12', 'TNFSF13', 'TNFRSF13B',\n",
    "       'TNFAIP1', 'TNFRSF11A', 'TNFAIP8L1', 'TNFSF9', 'TNFSF14',\n",
    "       'TNFRSF6B', 'TNFRSF13C'\n",
    "       ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata, \n",
    "                moi,\n",
    "                groupby = col_cell_type_fine,\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().savefig(f'{plots_path}/phenoAnalysis/thyAgeing_bSplit_cytokines_dotplot.pdf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- all: CCL5, CXCL10 (induced upon T cell contact and IFN, https://pubmed.ncbi.nlm.nih.gov/25476870/), CCL3 (positive and negative regulation of GCs by Tfr, https://www.sciencedirect.com/science/article/pii/0145212695000984)\n",
    "- B_med: CCL22, CCL17, IL6R (promotes plasma cell differentiation, https://www.sciencedirect.com/science/article/pii/0145212695000984), ILKAP, IL17RB, IL13RA1, IL4I1, TNFRSF11B, TNSF11 (GC-like less so)\n",
    "- TNFRSF18, TNFRSF4 (also more in plasma cells)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata, groupby = 'taa_l5', groups = ['B_transitional'], n_genes = 100, use_raw = False, method = 't-test', key_added = f'rank_genes_transitional', reference = 'rest', pts = False)\n",
    "\n",
    "sc.get.rank_genes_groups_df(adata, group = 'B_transitional', key = f'rank_genes_transitional').head(50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- FCRLA: B cell activation, in early-stage and GC B cells\n",
    "- CD37: mature B cells\n",
    "- CD1C: presentation of lipid-based antigens, marker of IgM memory cells (https://www.frontiersin.org/journals/immunology/articles/10.3389/fimmu.2021.602539/full)\n",
    "- VPREB3: early-stage and GC B cells\n",
    "- PTPRC: on all B cells, increases during differentiation\n",
    "- SYK: B cell differentiation (also follicular to GC, https://pmc.ncbi.nlm.nih.gov/articles/PMC4416743/), BCR signal amplification (https://elifesciences.org/articles/02069)\n",
    "- SMIM14: memory differntiation\n",
    "- RACK1: B cell differentation through stabilising PAX5 (https://www.nature.com/articles/s41423-024-01213-2)\n",
    "- CR2: complement activation, B cell maturation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.violin(adata ,keys = ['percent_mito', 'percent_ribo', 'n_counts', 'n_genes'], groupby = col_cell_type_fine)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct new anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_anno = adata.obs[['taa_l5']].astype(str).copy()\n",
    "anno_levels = pd.read_excel(f'{general_data_path}/curated/thyAgeing_full_curatedAnno_v9_2025-03-03_levels.xlsx')\n",
    "\n",
    "# Rename cell types\n",
    "ct_anno.loc[ct_anno['taa_l5'] == 'B_GC-like_explore','taa_l5'] = 'B_GC-like-locnt'\n",
    "ct_anno.loc[ct_anno['taa_l5'] == 'B_med','taa_l5'] = 'B_GC-like_AIRE+'\n",
    "ct_anno.loc[ct_anno['taa_l5'].isin(['B_transitional', 'B_mem_CR2+']),'taa_l5'] = 'B_mem_CR2+'\n",
    "\n",
    "ct_anno = ct_anno.reset_index(names='cell_id').merge(anno_levels, on = 'taa_l5', how = 'left').set_index('cell_id')[anno_levels.columns]\n",
    "\n",
    "ct_anno.to_csv(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}_curatedAnno_v6.csv')\n",
    "ct_anno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs.drop(columns = [c for c in ct_anno.columns if c in adata.obs.columns], inplace = True)\n",
    "adata.obs = adata.obs.join(ct_anno)\n",
    "\n",
    "with plt.rc_context(rc={'figure.figsize': calc_figsize(width = 'one-half', height = 120)}):\n",
    "    sc.pl.umap(adata, color = ct_anno.columns.tolist(), wspace = 0.3, return_fig = True, ncols = 2)\n",
    "    plt.savefig(f'{plots_path}/ctAnnotation/v6/thyAgeing_bSplit_scvi_{object_version}_ctAnnotation_v9.pdf', bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Marker expression\n",
    "b_markers = {'B_pan' : ['CD19', 'MS4A1'],\n",
    "             'B_naive' : ['CD22', 'SELL', 'IL4R', 'TCL1A', 'CR2', 'FOXO1', 'IGHM', 'IGHD',],\n",
    "               'B_mem' : ['CD27', 'CD38', 'FCRL4', 'FCRL5', 'CD44', 'PRDM1', 'IGHA1', 'IGHG1', 'IGHE'],\n",
    "               'B_mem_CD1C+' : ['CD1C', 'IGHM', 'FCRL3', 'CR2'],\n",
    "               'B_plasma' : ['PRDM1', 'XBP1', 'MZB1'], \n",
    "               'B_age-assoc' : ['TBX21', 'ITGAX',], # ITGAX = CD11c\n",
    "               'B_med' : ['HLA-DRA', 'AIRE', 'IL15', 'LTA', 'LTB'],\n",
    "               'DZ' : ['CXCR4', 'MYC', 'MKI67', 'TOP2A', 'AICDA', 'PCNA', 'BACH2'],\n",
    "               'LZ' : ['CXCR5', 'CD83', 'IRF4', 'CD86', 'MYBL1', 'SOCS3'],\n",
    "               'B_dev' : ['IGLL1', 'MME', 'RAG1', 'PAX5', 'EBF1', 'BCL11B'],\n",
    "               'B_dev_thy' : ['CD34', 'VPREB1', 'TYROBP',],}\n",
    "\n",
    "sc.pl.DotPlot(adata, \n",
    "                b_markers,\n",
    "                categories_order=[c for c in col_cell_type_fine_levels if c in adata.obs[col_cell_type_fine].tolist()],\n",
    "                groupby = col_cell_type_fine,\n",
    "                figsize = calc_figsize(width = 200, height = 50),\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().style(smallest_dot=1, largest_dot = 40).savefig(f'{plots_path}/ctAnnotation/v6/thyAgeing_bSplit_scvi_{object_version}_finalAnno_bMarkers_dotplot.pdf')  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot isotype usage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add BCR metadata\n",
    "bcr_meta = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_bSplit_scvi_{object_version}_bcr_{vdj_version}.csv', index_col=0)\n",
    "adata.obs = adata.obs.join(bcr_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bcr_df = adata.obs[['isotype', 'isotype_status','chain_status', 'locus_status', col_cell_type_broad, 'donor', 'sample', col_age_group]].copy()\n",
    "samples_to_exclude = bcr_df.groupby(['sample'])['isotype_status'].value_counts(normalize=True).where(lambda x: (x > 0.9) & (x.index.get_level_values(1) == 'No_contig')).dropna().index.get_level_values(0).tolist()\n",
    "bcr_df = bcr_df[~bcr_df['sample'].isin(samples_to_exclude)]\n",
    "\n",
    "bcr_df[['isotype_status','isotype']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = bcr_df.groupby(['sample', 'donor', col_age_group, col_cell_type_broad], observed = True)['isotype'].value_counts(normalize=True).reset_index(name='prop')\n",
    "df = df.groupby(['donor', col_age_group, col_cell_type_broad, 'isotype'], observed = True).agg(mean_prop = ('prop', 'mean')).reset_index()\n",
    "df = df.loc[(df['isotype'] != 'IgE') & (~df[col_cell_type_broad].isin(['B_dev', 'B_dev_thy']))]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotting.utils import calc_figsize, plot_faceted_grouped_boxplot\n",
    "\n",
    "p = plot_faceted_grouped_boxplot(data = df, x = 'isotype', y = 'mean_prop', hue = col_age_group, hue_order = ['infant', 'paed', 'adult'], order = df['isotype'].unique().tolist(), \n",
    "                             facet_kwargs = dict(col = col_cell_type_broad, col_wrap = 4, height = calc_figsize(width = 100, height = 30)[1], aspect = 1, col_order = df[col_cell_type_broad].unique().tolist()),\n",
    "                             add_stats = True, #save_stats = f'{data_path}/analyses/vdj/cdr3Analysis/thyAgeing_devSplit_TRB_kidera_prod_', \n",
    "                             format_percent = True, format_log = False, x_label = 'Cell population', y_label = 'Mean TRB Kidera Factor', legend_title='Age group', figsize = calc_figsize(width = 100, height=50),\n",
    "                             ylim = None, #legend_kwargs = dict(loc = \"upper left\", bbox_to_anchor=(1, 1))\n",
    "                             )\n",
    "p.tight_layout()\n",
    "#plt.savefig(f'{plots_path}/vdj/cdr3Analysis/thyAgeing_newTSplit_TRB_kidera_prod.pdf', bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[col_cell_type_fine].unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "bcr_df['isotype_short'] = bcr_df['isotype'].apply(lambda x : 'Multiple' if (isinstance(x, str) and '|' in x) else ('Orphan' if isinstance(x, float) else x))\n",
    "#bcr_df = bcr_df.loc[bcr_df['isotype_short'] != 'No_contig']\n",
    "\n",
    "isotype_freq = bcr_df.groupby(['sample', 'donor', col_age_group])['isotype_short'].value_counts(normalize=True).reset_index(name = 'prop')\n",
    "isotype_freq = isotype_freq.groupby(['donor', col_age_group, 'isotype_short']).agg(mean_prop = ('prop', 'mean'),\n",
    "                                                                                 sd_prop = ('prop', 'std')).reset_index()\n",
    "\n",
    "isotype_freq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i isotype_freq -w 300 -h 150 -u mm\n",
    "\n",
    "plot_data = isotype_freq %>%\n",
    "tidyr::drop_na(mean_prop) %>%\n",
    "dplyr::mutate(!!col_age_group := factor(!!rlang::sym(col_age_group), levels = col_age_group_levels))\n",
    "\n",
    "kruskal_test_res = plot_data %>%\n",
    "dplyr::group_by(isotype_short) %>%\n",
    "dplyr::filter(length(unique(!!rlang::sym(col_age_group))) > 1) %>%\n",
    "rstatix::group_by(isotype_short) %>%\n",
    "rstatix::kruskal_test(as.formula(paste0('mean_prop ~', col_age_group))) %>%\n",
    "rstatix::adjust_pvalue(method = 'BH') %>%\n",
    "rstatix::add_significance() \n",
    "kruskal_test_res %>% readr::write_csv(file.path(data_path, 'analysis/vdj', 'thyAgeing_isotypeFreq_byAge_kruskal.csv'))\n",
    "\n",
    "dunn_test_res = plot_data %>%\n",
    "dplyr::group_by(isotype_short) %>%\n",
    "dplyr::filter(length(unique(!!rlang::sym(col_age_group))) > 1) %>%\n",
    "rstatix::group_by(isotype_short) %>%\n",
    "rstatix::dunn_test(as.formula(paste0('mean_prop ~', col_age_group)), p.adjust.method = 'BH') %>%\n",
    "rstatix::add_significance() %>%\n",
    "rstatix::add_xy_position(x = 'isotype_short', step.increase = 0.05)\n",
    "dunn_test_res %>% readr::write_csv(file.path(data_path, 'analysis/vdj', 'thyAgeing_isotypeFreq_byAge_dunn.csv'))\n",
    "\n",
    "isotype_freq %>%\n",
    "rstatix::group_by(!!rlang::sym(col_age_group), isotype_short) %>%\n",
    "rstatix::get_summary_stats(mean_prop) %>%\n",
    "readr::write_csv(file.path(data_path, 'analysis/vdj', 'thyAgeing_isotypeFreq_byAge_summary.csv'))\n",
    "\n",
    "plot_data %>%\n",
    "ggplot(aes(x = isotype_short, y = mean_prop, color = !!rlang::sym(col_age_group))) +\n",
    "geom_boxplot(outlier.size = 0.5, position = position_dodge2(preserve = 'single')) +\n",
    "ggpubr::stat_pvalue_manual(dunn_test_res, hide.ns = T, label = 'p.adj.signif', tip.length = 0.01) +\n",
    "#ggforce::facet_row(~taa_l0, scales = 'free', space = 'free') +\n",
    "labs(x = 'Cell population', y = \"Mean frequency\", color = 'Age group') +\n",
    "grafify::scale_color_grafify(palette = \"fishy\") +\n",
    "scale_y_continuous(labels = scales::percent_format(scale = 100)) +\n",
    "theme_simple(facet = F) +\n",
    "theme(axis.text.x = element_text(angle = 90, hjust = 1))\n",
    "ggsave(paste0(\"thyAgeing_isotypeFreq_byAge_boxplot.png\"), path = file.path(plots_path, 'vdjAnalysis'), \n",
    "       width = 150, height = 100, units = \"mm\", dpi = 300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clonotypes by donor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clone numbers\n",
    "df = bcr_df.groupby(['donor', col_age_group])['changeo_clone_id'].nunique().reset_index(name = 'n_clones')\n",
    "df = df.merge(bcr_df.groupby(['donor', col_age_group]).size().reset_index(name = 'n_cells'))\n",
    "df['n_clones_norm'] = df['n_clones'] / df['n_cells']\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clone size\n",
    "df = bcr_df.groupby(['donor', col_age_group,'changeo_clone_id'], observed=True).size().reset_index(name = 'clone_size')\n",
    "donor_order = bcr_df[['donor', 'age_months']].drop_duplicates().dropna().sort_values('age_months')['donor'].tolist()\n",
    "\n",
    "with sns.plotting_context('paper', font_scale = 1.4):\n",
    "    p = sns.violinplot(data = df, x = 'donor', y = 'clone_size', hue = col_age_group, hue_order = ['infant', 'paed', 'adult'], palette = 'colorblind',\n",
    "                       density_norm = 'width', aspect = 1.3, dodge = False, order = donor_order)\n",
    "    p.xaxis.set_tick_params(rotation=45)\n",
    "    p.set_xlabel('Donor')\n",
    "    p.set_ylabel('Clone size')\n",
    "    p.legend(title = 'Age group')\n",
    "    p.set_yscale('log') \n",
    "    \n",
    "plt.savefig(f'{plots_path}/vdjAnalysis/thyAgeing_cloneSize_byAge_violin.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lineage analysis using Immcantation\n",
    "\n",
    "Use downser from immcancantation (https://dowser.readthedocs.io/en/latest/vignettes/Building-Trees-Vignette/) to construct lineage trees "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i data_path,vdj_version\n",
    "\n",
    "# Load bcr data\n",
    "bcr = read_tsv(file.path(data_path, paste0('/analysis/vdj/hypermutation/thymusAgeing_bcrFiltered_', vdj_version, '_airr.tsv')))\n",
    "\n",
    "# Remove any cells without a cell type and with stop codons\n",
    "bcr = bcr %>% filter(!is.na(!!rlang::sym(col_cell_type_fine)),\n",
    "                      !stop_codon)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Remove cells with multiple heavy chains\n",
    "multi_heavy <- table(dplyr::filter(bcr, locus == \"IGH\")$cell_id)\n",
    "multi_heavy_cells <- names(multi_heavy)[multi_heavy > 1]\n",
    "\n",
    "print(paste('Removing', length(multi_heavy_cells), 'cells with multiple heavy chains'))\n",
    "\n",
    "bcr = bcr %>%\n",
    "dplyr::filter(!cell_id %in% multi_heavy_cells)\n",
    "\n",
    "nrow(bcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Remove cells without any heavy chain\n",
    "heavy_cells <- dplyr::filter(bcr, locus == \"IGH\")$cell_id\n",
    "light_cells <- dplyr::filter(bcr, locus == \"IGK\" | locus == \"IGL\")$cell_id\n",
    "no_heavy_cells <- light_cells[which(!light_cells %in% heavy_cells)]\n",
    "\n",
    "print(paste('Removing', length(no_heavy_cells), 'cells without heavy chain'))\n",
    "\n",
    "bcr = bcr %>%\n",
    "dplyr::filter(!cell_id %in% no_heavy_cells)\n",
    "\n",
    "nrow(bcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "bcr %>%\n",
    "rstatix::freq_table(donor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "# Calculate distribution of distance to nearest\n",
    "dist_nearest <- shazam::distToNearest(bcr,\n",
    "                                      cellIdColumn = 'cell_id', # Single-cell mode\n",
    "                                      fields = 'donor',\n",
    "                                      locusColumn = 'locus',\n",
    "                                      onlyHeavy=FALSE,\n",
    "                                      nproc = 4)\n",
    "\n",
    "# find threshold for cloning automatically and initialize the Gaussian fit\n",
    "# parameters of the nearest-neighbor\n",
    "\n",
    "# distance of clones using distribution of distance to nearest\n",
    "threshold_output <- shazam::findThreshold(dist_nearest$dist_nearest,\n",
    "                                          method = \"gmm\", model = \"gamma-norm\", \n",
    "                                          cutoff = \"user\", spc = 0.95)\n",
    "threshold <- threshold_output@threshold\n",
    "threshold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Cluster clones\n",
    "results <- scoper::hierarchicalClones(bcr,\n",
    "                              cell_id = \"cell_id\",\n",
    "                              clone = 'clone_id',\n",
    "                              threshold = threshold,\n",
    "                              only_heavy = FALSE, split_light = TRUE,\n",
    "                              summarize_clones = FALSE,\n",
    "                              fields = \"donor\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "results %>% colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# calculate and plot the rank-abundance curve\n",
    "abund <- alakazam::estimateAbundance(dplyr::filter(results, locus == \"IGH\"),\n",
    "                           group = \"donor\", nboot = 100)\n",
    "\n",
    "abund_plot <- alakazam::plot(abund, silent=T)\n",
    "abund_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i plots_path -w 200 -h 200 -u mm -o clone_sizes\n",
    "\n",
    "# get clone sizes using dplyr functions\n",
    "clone_sizes <- alakazam::countClones(results,\n",
    "                           groups = \"donor\")\n",
    "\n",
    "# plot cells per clone\n",
    "ggplot(clone_sizes, aes(x = seq_count)) +\n",
    "  geom_bar() +\n",
    "  facet_wrap(~donor, scales = 'free_y') +\n",
    "  labs(x = \"Sequences per clone\", x = 'Log10(Clones)') +\n",
    "  scale_y_continuous(expand = expansion(mult = c(0, 0.05)), trans = 'log10') +\n",
    "  theme_bw()\n",
    "ggsave('thyAgeing_bSplit_seqPerClone.png',path = file.path(plots_path, 'vdjAnalysis', 'clonotypes'),\n",
    "       height = 200, width = 200, units = 'mm', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# Remove sequences assigned to no clones\n",
    "results <- dplyr::filter(results, !is.na(clone_id))\n",
    "\n",
    "# Read in IMGT reference sequences\n",
    "references <- dowser::readIMGT(dir = \"/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/B_compartment/data/references/immcantation_imgt/human/vdj\")\n",
    "\n",
    "# Reconstruct germlines\n",
    "results <- dowser::createGermlines(results, references, fields =c(\"donor\"), nproc = 4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "results %>% colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "clones <- dowser::formatClones(results,\n",
    "                       traits = c(col_cell_type_fine),\n",
    "                       columns = c('donor'),\n",
    "                       cell = 'cell_id',\n",
    "                       #text_fields = c(\"taa_l5\"), \n",
    "                       minseq = 3, nproc = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "clones %>% colnames()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "Sys.setenv(IGPHYML_PATH = \"/nfs/team205/lm25/condaEnvs/thymusAgeing/share/igphyml/motifs\")\n",
    "\n",
    "trees <- dowser::getTrees(clones, nproc = 8, build=\"igphyml\", exec=\"/nfs/team205/lm25/condaEnvs/thymusAgeing/bin/igphyml\")\n",
    "\n",
    "trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o trees_df\n",
    "\n",
    "trees_df = trees %>%\n",
    "dplyr::select(clone_id, locus, seqs, donor) %>%\n",
    "data.frame() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = trees_df.merge(latest_meta[['donor', col_age_group]].drop_duplicates(), on = 'donor')\n",
    "df = df.groupby(['donor', col_age_group]).agg(n_clones = ('clone_id', 'nunique')).reset_index()\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i donor_meta -h 100 -w 200 -u mm\n",
    "\n",
    "trees %>%\n",
    "rstatix::freq_table(donor) %>%\n",
    "dplyr::right_join(donor_meta, by = c('donor' = 'donor')) %>%\n",
    "dplyr::filter(donor %in% bcr$donor) %>%\n",
    "tidyr::replace_na(list(n = 0, prop = 0)) %>%\n",
    "dplyr::mutate(age_group = factor(age_group, levels = c('infant', 'paed', 'adult'))) %>%\n",
    "ggplot(aes(x = donor, y = n, fill = age_group)) +\n",
    "  geom_bar(stat = 'identity') +\n",
    "  labs(x = 'Age group', y = 'Count') +\n",
    "  ggforce::facet_row(~ age_group, scale = 'free_x', space = 'free') +\n",
    "  scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +\n",
    "  grafify::scale_fill_grafify(palette = \"fishy\") +\n",
    "  labs(fill = 'Age group', x = 'Donor', y = 'N(trees)') +\n",
    "  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +\n",
    "  theme_simple(facet = T)\n",
    "  \n",
    "ggsave('thyAgeing_bSplit_nTreesByDonor_barplot.png',path = file.path(plots_path, 'vdjAnalysis', 'clonotypes'),\n",
    "       height = 100, width = 200, units = 'mm', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i vdj_version\n",
    "\n",
    "saveRDS(trees, file = file.path(data_path, paste0('objects/vdj/thymusAgeing_bcrFiltered_', vdj_version, '_trees.rds')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "tree_plots <- dowser::plotTrees(trees, tips = \"taa_l5\", tipsize = 2)\n",
    "\n",
    "dowser::treesToPDF(tree_plots, file.path(plots_path, 'vdjAnalysis', 'clonotypes', 'thyAgeing_bSplit_bLineageTrees.pdf'), nrow = 2, ncol = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "# calculate switches along trees compared to 100 random permutations \n",
    "# this may take a while, and can be parallelized using nproc\n",
    "switches = dowser::findSwitches(clones, permutations=100, trait='donor', \n",
    "  igphyml=\"/nfs/team205/lm25/condaEnvs/thymusAgeing/bin/igphyml\", fixtrees=TRUE)\n",
    "\n",
    "ps = dowser::testPS(switches$switches)\n",
    "\n",
    "print(ps$means)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "print(ps$means)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
