{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Update T/NK annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import session_info\n",
    "from datetime import datetime\n",
    "today = datetime.today().strftime('%Y-%m-%d')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import hdf5plugin\n",
    "\n",
    "# Add repo path to sys path (allows to access scripts and metadata from repo)\n",
    "repo_path = '/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/T_NK_compartment'\n",
    "sys.path.insert(1, repo_path) \n",
    "sys.path.insert(2, '/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts')\n",
    "\n",
    "# Autoreload custom scripts\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# Define paths\n",
    "plots_path = f'{repo_path}/plots/'\n",
    "data_path = f'{repo_path}/data/'\n",
    "model_path = os.path.join(repo_path, 'models')\n",
    "general_data_path = '/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/data'\n",
    "\n",
    "print('Dir for plots: {}'.format(plots_path))\n",
    "print('Dir for data: {}'.format(data_path))\n",
    "\n",
    "# Formatting\n",
    "# from matplotlib import font_manager\n",
    "# font_manager.fontManager.addfont(\"/nfs/team205/ny1/ThymusSpatialAtlas/software/Arial.ttf\")\n",
    "# plt.style.use('/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts/plotting/thyAgeing.mplstyle')\n",
    "\n",
    "# Import custom scripts\n",
    "from utils import get_latest_version,update_obs,freq_by_donor\n",
    "from anno_levels import get_ct_levels, get_ct_palette, age_group_levels, age_group_palette\n",
    "from plotting.utils import plot_grouped_boxplot, calc_figsize"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## B_dev_thy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load old T annotations\n",
    "t_anno = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_v8_2024-11-07_curatedAnno_v6.csv', index_col = 0)\n",
    "\n",
    "# Load B annotations\n",
    "from anno_levels import b_path\n",
    "b_anno = pd.read_csv(f'{b_path}/data/preprocessing/ctAnnotation/thyAgeing_bSplitxTissue_scvi_v2_2025-02-20_v5.csv', index_col = 0)\n",
    "b_barcodes = np.intersect1d(b_anno.index, t_anno.index)\n",
    "\n",
    "# Load anno levels\n",
    "anno_level_sheet = pd.read_excel(f'{general_data_path}/curated/thyAgeing_full_curatedAnno_v8_2025-03-03_levels.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_level_sheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Construct new annotations\n",
    "new_anno = t_anno[['taa_l5']].copy()\n",
    "new_anno.loc[new_anno['taa_l5'] == 'B_dev_thy', 'taa_l5'] = pd.NA\n",
    "new_anno.loc[b_barcodes, 'taa_l5'] = 'B_dev_thy'\n",
    "\n",
    "new_anno = new_anno.reset_index(names='names').merge(anno_level_sheet, left_on = 'taa_l5', right_on = 'taa_l5', how = 'left').set_index('names')\n",
    "new_anno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_anno['taa_l5'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check number of NAs\n",
    "new_anno['taa_l5'].isna().sum() # Doublet B_dev_thy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save new annotations\n",
    "new_anno.to_csv(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_v8_2024-11-07_curatedAnno_v7.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DN compartment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata\n",
    "object_version = 'v4_2025-02-04'\n",
    "adata = ad.read_h5ad(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_{object_version}.zarr', backed = 'r')\n",
    "\n",
    "# Add latest annotations to adata\n",
    "ct_anno = pd.read_csv(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_{object_version}_curatedAnno_v8.csv', index_col = 0)\n",
    "for c in ct_anno.columns:\n",
    "    if c in adata.obs.columns:\n",
    "        adata.obs.drop(c, axis = 1, inplace = True)\n",
    "adata.obs = adata.obs.join(ct_anno)\n",
    "\n",
    "adata = adata[adata.obs['taa_l4'].isin(['T_DN(early)' ,'T_DN(P)', 'T_DN(Q)', 'T_DN(late)', 'B_dev_thy'])].to_memory()\n",
    "\n",
    "adata.obs['taa_l5'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load leiden clustr 9\n",
    "adata_c9 = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_v7_2024-11-06.zarr', backed= 'r')\n",
    "\n",
    "leiden_c9 = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_v7_2024-11-06_leidenClusters.csv', index_col = 0)\n",
    "adata_c9.obs = adata_c9.obs.join(leiden_c9)\n",
    "\n",
    "adata_c9 = adata_c9[adata_c9.obs['leiden_r2.5'] == 9].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata_c9, \n",
    "              var_names = ct_markers_filtered,\n",
    "              groupby = 'chemistry_simple',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_c9.obs['taa_l5'] = 'T_DN(Q)'\n",
    "adata_concat = adata.concatenate(adata_c9, index_unique = None).copy()\n",
    "\n",
    "adata_concat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adata_concat = adata_concat[adata_concat.obs['scrublet_score'] <= 0.4]\n",
    "# adata_concat = adata_concat[adata_concat.obs['percent_mito'] < .08]\n",
    "# adata_concat = adata_concat[adata_concat.obs['percent_ribo'] < .7]\n",
    "\n",
    "# adata_concat.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check for duplicates\n",
    "adata_concat.obs_names.duplicated().sum() # Check for duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check whether X is raw counts\n",
    "adata_concat.X[:100, :100].sum() % 1 == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load curated markers and make dictionary\n",
    "ct_markers_df = pd.read_excel(f'{data_path}/curated/curatedCellMarkers_LMM.xlsx', sheet_name = 'Final')\n",
    "ct_markers = ct_markers_df.groupby('cell_label_fine')['marker_gene'].agg(list).to_dict()\n",
    "\n",
    "# Filter markers of interest\n",
    "ctoi = ['B_dev','T_DN(early)', 'T_DN(P)', 'T_DN(Q)','T_DP(P)']\n",
    "ct_markers_filtered = {c: [g for g in ct_markers[c] if g in adata.var_names] for c in ctoi if c in ct_markers}\n",
    "ct_markers_filtered['B_dev'].append('IFITM3')\n",
    "ct_markers_filtered['T_DN(early)'].append('SPINK2')\n",
    "ct_markers_filtered['T_DN(early)'].append('NOTCH1')\n",
    "\n",
    "import pprint\n",
    "pprint.pprint(ct_markers_filtered, compact=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat.layers['counts'] = adata_concat.X.copy()\n",
    "sc.pp.normalize_total(adata_concat, target_sum=1e4)\n",
    "sc.pp.log1p(adata_concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata_concat, \n",
    "              var_names = ct_markers_filtered,\n",
    "              categories_order = ['B_dev_thy', 'T_ETP', 'T_DN(early)', 'T_DN(P)', 'T_DN(Q)', 'T_DN(late)'],\n",
    "              groupby = 'taa_l5',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat.obs[['taa_l5']].to_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_dnSplit_all_scvi_{object_version}_tempAnno.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Score cell cycle\n",
    "cell_cycle_genes_df = pd.read_excel('/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/T_NK_compartment/data/curated/regev_cell_cycle_genes.xlsx', skiprows = 6, usecols=[0,1])\n",
    "\n",
    "cell_cycle_genes = {}\n",
    "for col in cell_cycle_genes_df.columns:\n",
    "    cell_cycle_genes_df[col] = cell_cycle_genes_df[col].str.strip()\n",
    "    cell_cycle_genes[col] = [g.strip() for g in cell_cycle_genes_df[col].dropna().tolist() if g.strip() in adata.var_names]\n",
    "    \n",
    "import pprint\n",
    "pprint.pprint(cell_cycle_genes, compact=True)\n",
    "\n",
    "sc.tl.score_genes_cell_cycle(adata_concat, s_genes=cell_cycle_genes['G1/S'], g2m_genes=cell_cycle_genes['G2/M'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_cycle_score_df = adata_concat.obs[['S_score', 'G2M_score', 'donor', 'age_group', 'taa_l5', 'phase', 'study']].copy()\n",
    "#cell_cycle_score_df = cell_cycle_score_df.join(knn_anno)\n",
    "df = cell_cycle_score_df.groupby(['age_group', 'taa_l5', 'donor', 'study'], observed = True)['phase'].value_counts(normalize = True).unstack().fillna(0).reset_index()\n",
    "df = df.melt(id_vars = ['age_group', 'taa_l5', 'donor', 'study'], value_name = 'prop', var_name = 'phase')\n",
    "df['taa_l5'] = pd.Categorical(df['taa_l5'], categories = ['T_ETP','T_DN(early)', 'T_DN(P)', 'T_DN(Q)', 'T_DN(late)', 'B_dev_thy'], ordered=True)\n",
    "df['taa_l5'] = df['taa_l5'].cat.remove_unused_categories()\n",
    "\n",
    "sns.catplot(data=df, x='taa_l5', y='prop', hue = 'age_group', hue_order = ['infant', 'paed', 'adult', 'geriatric'], kind=\"box\", palette = 'colorblind',\n",
    "            col='phase', col_order=['G1', 'S','G2M'], \n",
    "            #facet_kws={'sharey': False, 'sharex': True}, \n",
    "            height=2, aspect=2, col_wrap=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scvi_wrapper import run_scvi\n",
    "object_version = f'v3_{today}'\n",
    "\n",
    "# Run scvi\n",
    "scvi_run = run_scvi(adata_concat, \n",
    "                    layer_raw = 'X', \n",
    "                    # Excluded genes\n",
    "                    include_genes=[g for g in ct_markers_filtered.values() for g in g], \n",
    "                    exclude_cc_genes=False, exclude_mt_genes=True, \n",
    "                    exclude_vdjgenes = True, remove_cite = False,\n",
    "                    # Highly variable gene selection\n",
    "                    batch_hv=\"study\", hvg = 200, #span = 1,\n",
    "                    hvg_selection = 'experimental',\n",
    "                    # scVI \n",
    "                    batch_scvi=\"sample\",\n",
    "                    cat_cov_scvi=[\"donor\", \"chemistry_simple\", \"sex\"], \n",
    "                    cont_cov_scvi=[], \n",
    "                    max_epochs=400, batch_size=2000, early_stopping = True, early_stopping_patience = 15, early_stopping_min_delta = 10.0,\n",
    "                    plan_kwargs = {'lr': 0.001, 'reduce_lr_on_plateau' : True, 'lr_patience' : 10, 'lr_threshold' : 20}, \n",
    "                    n_layers = 3, n_latent = 30, dispersion = 'gene-batch',\n",
    "                    # Leiden clustering\n",
    "                    leiden_clustering = None, col_cell_type = ['taa_l5'], \n",
    "                    fig_dir = f'{plots_path}/preprocessing/scvi', fig_prefix = f'thyAgeing_dnSplit_scvi_{object_version}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save adata and scvi model\n",
    "overwrite = True\n",
    "\n",
    "anno_cols = [c for c in scvi_run['data'].obs.columns if '_pred_' in c or '_prob_' in c or 'taa' in c or 'anno_' in c or 'leiden' in c]\n",
    "if not os.path.exists(f'{data_path}/thyAgeing_dnSplit_scvi_{object_version}.zarr') or overwrite:\n",
    "    print('Saving new adata version: {}'.format(object_version))\n",
    "    scvi_run['data'].obs = scvi_run['data'].obs.drop(columns=anno_cols)\n",
    "    scvi_run['data'].write_h5ad(\n",
    "        f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}.zarr',\n",
    "        compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "        compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,\n",
    "    )\n",
    "    scvi_run['vae'].save(f'{model_path}/thyAgeing_dnSplit_scvi_{object_version}', save_anndata=False, overwrite=overwrite)\n",
    "else:\n",
    "    print('File already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}.zarr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Louvain clustering\n",
    "res_list = [1.5,2.0]\n",
    "for res in res_list:\n",
    "    sc.tl.leiden(adata, resolution = res, key_added = f\"leiden_r{res}\")\n",
    "adata.obs[[f'leiden_r{str(r)}' for r in res_list]] = adata.obs[[f'leiden_r{str(r)}' for r in res_list]].astype('category')\n",
    "\n",
    "adata.obs[[f'leiden_r{str(r)}' for r in res_list]].to_csv(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}_leidenClusters.csv')\n",
    "\n",
    "sc.pl.umap(adata, color = ['leiden_r1.5', 'leiden_r2.0'], show = False, return_fig = True)\n",
    "plt.savefig(f'{plots_path}/preprocessing/scvi/thyAgeing_dnSplit_scvi_{object_version}_leidenClusters.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Marker expression and ct assignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}.zarr')\n",
    "\n",
    "leiden_clustering = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}_leidenClusters.csv', index_col = 0)\n",
    "adata.obs = adata.obs.join(leiden_clustering)\n",
    "adata.obs[leiden_clustering.columns] = adata.obs[leiden_clustering.columns].astype(int).astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata, \n",
    "                ct_markers_filtered,\n",
    "                groupby = 'leiden_r2.0',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_dnMarkers_leiden2.0_dotplot.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check QC metrics\n",
    "sc.pl.violin(adata, keys = ['percent_mito', 'percent_ribo', 'n_counts', 'n_genes', 'scrublet_score'], groupby='leiden_r2.0', show = False)\n",
    "plt.savefig(f'{plots_path}/preprocessing/scvi/thyAgeing_dnSplit_scvi_{object_version}_leiden2.0_qc.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove clusters:\n",
    "- high mito: 16\n",
    "- high scrublet score: 40"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dn_cluster_assignments = {'B_dev_thy' : [28,25,37],\n",
    "                          'T_DN(early)' : [10,39,20,11],\n",
    "                          'T_DN(P)' : [3,6,21,23,33,39,],\n",
    "                          'T_DN(Q)' : [0,1,2,4,5,7,8,12,13,14,15,16,17,18,19,22,24,27,29,30,31,32,36,38],\n",
    "                          'T_DN(late)' : [2,9,26,34,35,40],\n",
    "                          'Remove' : [16,40]}\n",
    "\n",
    "np.array([c for c in adata.obs['leiden_r2.0'].unique().tolist() if c not in [c for c in dn_cluster_assignments.values() for c in c]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = ['CD3D', 'VPREB1', 'CD34', 'RAG1', 'CD8A'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['temp_anno'] = pd.NA\n",
    "#dn_cluster_assignments = {k: [c for c in dn_cluster_assignments[k]] for k in dn_cluster_assignments.keys()}\n",
    "for k, v in dn_cluster_assignments.items():\n",
    "    adata.obs.loc[adata.obs['leiden_r2.0'].isin(v), 'temp_anno'] = k\n",
    "    \n",
    "sc.pl.umap(adata, color = 'temp_anno', return_fig = True)\n",
    "plt.savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_tempAnno_umap.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata[adata.obs['temp_anno'] != 'Remove'], \n",
    "                ct_markers_filtered,\n",
    "                groupby = 'temp_anno',\n",
    "                categories_order= ['B_dev_thy', 'T_DN(early)', 'T_DN(P)', 'T_DN(Q)', 'T_DN(late)'],\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_dnMarkers_tempAnno_dotplot.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[adata.obs['temp_anno'] == 'B_dev_thy']['donor'].value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[['temp_anno']].to_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_tempAnno.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Score cell cycle\n",
    "cell_cycle_genes_df = pd.read_excel('/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/T_NK_compartment/data/curated/regev_cell_cycle_genes.xlsx', skiprows = 6, usecols=[0,1])\n",
    "\n",
    "cell_cycle_genes = {}\n",
    "for col in cell_cycle_genes_df.columns:\n",
    "    cell_cycle_genes_df[col] = cell_cycle_genes_df[col].str.strip()\n",
    "    cell_cycle_genes[col] = [g.strip() for g in cell_cycle_genes_df[col].dropna().tolist() if g.strip() in adata.var_names]\n",
    "    \n",
    "import pprint\n",
    "pprint.pprint(cell_cycle_genes, compact=True)\n",
    "\n",
    "sc.tl.score_genes_cell_cycle(adata_dn, s_genes=cell_cycle_genes['G1/S'], g2m_genes=cell_cycle_genes['G2/M'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_cycle_score_df = adata_dn.obs[['S_score', 'G2M_score', 'donor', 'age_group', 'temp_anno', 'phase', 'study']].copy()\n",
    "#cell_cycle_score_df = cell_cycle_score_df.join(knn_anno)\n",
    "df = cell_cycle_score_df.groupby(['age_group', 'temp_anno', 'donor', 'study'])['phase'].value_counts(normalize = True).unstack().fillna(0).reset_index()\n",
    "df = df.melt(id_vars = ['age_group', 'temp_anno', 'donor', 'study'], value_name = 'prop', var_name = 'phase')\n",
    "df['temp_anno'] = pd.Categorical(df['temp_anno'], categories = ['T_ETP','T_DN(early)', 'T_DN(P)', 'T_DN(Q)', 'T_DN(late)', 'T_B_dev'], ordered=True)\n",
    "df['temp_anno'] = df['temp_anno'].cat.remove_unused_categories()\n",
    "\n",
    "sns.catplot(data=df, x='temp_anno', y='prop', hue = 'age_group', hue_order = ['infant', 'paed', 'adult', 'geriatric'], kind=\"box\", palette = 'colorblind',\n",
    "            col='phase', col_order=['G1', 'S','G2M'], \n",
    "            facet_kws={'sharey': False, 'sharex': True}, \n",
    "            height=2, aspect=2, col_wrap=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NEw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scvi_wrapper import run_scvi\n",
    "object_version = f'v2_{today}'\n",
    "\n",
    "# Run scvi\n",
    "scvi_run = run_scvi(adata_concat, \n",
    "                    layer_raw = 'X', \n",
    "                    # Excluded genes\n",
    "                    include_genes=[], exclude_cc_genes=False, exclude_mt_genes=True, \n",
    "                    exclude_vdjgenes = True, remove_cite = False,\n",
    "                    # Highly variable gene selection\n",
    "                    batch_hv=\"study\", hvg = 200, #span = 1,\n",
    "                    hvg_selection = 'experimental',\n",
    "                    # scVI \n",
    "                    batch_scvi=\"sample\",\n",
    "                    cat_cov_scvi=[\"donor\", \"chemistry_simple\", \"sex\"], \n",
    "                    cont_cov_scvi=[], \n",
    "                    max_epochs=400, batch_size=2000, early_stopping = True, early_stopping_patience = 15, early_stopping_min_delta = 10.0,\n",
    "                    plan_kwargs = {'lr': 0.001, 'reduce_lr_on_plateau' : True, 'lr_patience' : 10, 'lr_threshold' : 20}, \n",
    "                    n_layers = 3, n_latent = 30, dispersion = 'gene-batch',\n",
    "                    # Leiden clustering\n",
    "                    leiden_clustering = None, col_cell_type = ['taa_l5'], \n",
    "                    fig_dir = f'{plots_path}/preprocessing/scvi', fig_prefix = f'thyAgeing_dnSplit_scvi_{object_version}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save adata and scvi model\n",
    "overwrite = True\n",
    "\n",
    "anno_cols = [c for c in scvi_run['data'].obs.columns if '_pred_' in c or '_prob_' in c or 'taa' in c or 'anno_' in c or 'leiden' in c]\n",
    "if not os.path.exists(f'{data_path}/thyAgeing_dnSplit_scvi_{object_version}.zarr') or overwrite:\n",
    "    scvi_run['data'].obs = scvi_run['data'].obs.drop(columns=anno_cols)\n",
    "    scvi_run['data'].write_h5ad(\n",
    "        f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}.zarr',\n",
    "        compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "        compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,\n",
    "    )\n",
    "    scvi_run['vae'].save(f'{model_path}/thyAgeing_dnSplit_scvi_{object_version}', save_anndata=False, overwrite=overwrite)\n",
    "else:\n",
    "    print('File already exists')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "object_version = 'v2_2025-04-04'\n",
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}.zarr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Louvain clustering\n",
    "res_list = [2.0,2.5]\n",
    "for res in res_list:\n",
    "    sc.tl.leiden(adata, resolution = res, key_added = f\"leiden_r{res}\")\n",
    "adata.obs[[f'leiden_r{str(r)}' for r in res_list]] = adata.obs[[f'leiden_r{str(r)}' for r in res_list]].astype('category')\n",
    "\n",
    "adata.obs[[f'leiden_r{str(r)}' for r in res_list]].to_csv(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}_leidenClusters.csv')\n",
    "\n",
    "sc.pl.umap(adata, color = ['leiden_r2.5', 'leiden_r2.0'], show = False, return_fig = True)\n",
    "plt.savefig(f'{plots_path}/preprocessing/scvi/thyAgeing_dnSplit_scvi_{object_version}_leidenClusters.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = ['leiden_r2.5', 'leiden_r2.0'], show = False, return_fig = True, ncols = 1)\n",
    "plt.savefig(f'{plots_path}/preprocessing/scvi/thyAgeing_dnSplit_scvi_{object_version}_leidenClusters.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}.zarr')\n",
    "\n",
    "leiden_clustering = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_dnSplit_scvi_{object_version}_leidenClusters.csv', index_col = 0)\n",
    "adata.obs = adata.obs.join(leiden_clustering)\n",
    "adata.obs[leiden_clustering.columns] = adata.obs[leiden_clustering.columns].astype(int).astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load curated markers and make dictionary\n",
    "ct_markers_df = pd.read_excel(f'{data_path}/curated/curatedCellMarkers_LMM.xlsx', sheet_name = 'Final')\n",
    "ct_markers = ct_markers_df.groupby('cell_label_fine')['marker_gene'].agg(list).to_dict()\n",
    "\n",
    "import pprint\n",
    "pprint.pprint(ct_markers, compact=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter markers of interest\n",
    "ctoi = ['B_dev','T_DN(early)', 'T_DN(P)', 'T_DN(Q)','T_DP(P)']\n",
    "ct_markers_filtered = {c: [g for g in ct_markers[c] if g in adata.var_names] for c in ctoi if c in ct_markers}\n",
    "ct_markers_filtered['B_dev'].append('IFITM3')\n",
    "ct_markers_filtered['T_DN(early)'].append('SPINK2')\n",
    "ct_markers_filtered['T_DN(early)'].append('NOTCH1')\n",
    "sc.pl.DotPlot(adata, \n",
    "                ct_markers_filtered,\n",
    "                groupby = 'leiden_r2.5',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_dnMarkers_leiden2.5_dotplot.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check QC metrics\n",
    "sc.pl.violin(adata, keys = ['percent_mito', 'percent_ribo', 'n_counts', 'n_genes', 'scrublet_score'], groupby='leiden_r2.5', show = False)\n",
    "plt.savefig(f'{plots_path}/preprocessing/scvi/thyAgeing_dnSplit_scvi_{object_version}_leiden2.5_qc.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Leiden 2.5\n",
    "dn_cluster_assignments = {'B_dev_thy' : [7,32,44,36],\n",
    "                          'T_DN(early)' : [15,28,34],\n",
    "                          'T_DN(P)' : [0,10,12,29,27,35,37,41],\n",
    "                          'T_DN(Q)' : [1,2,4,5,6,8,9,11,13,14,16,17,18,38,39,20,21,22,23,24,25,26,38,39,42,43,46,47,31,33,30],\n",
    "                          'T_DN(late)' : [3,19],\n",
    "                          'Remove' : [45,21,40]} # LowQC and doublets\n",
    "\n",
    "np.array([c for c in adata.obs['leiden_r2.5'].unique().tolist() if c not in [c for c in dn_cluster_assignments.values() for c in c]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Leiden 2.0\n",
    "dn_cluster_assignments = {'B_dev_thy' : [11,17],\n",
    "                          'T_DN(early)' : [5],\n",
    "                          'T_DN(P)' : [1,6,35,33,34,28,8,24],\n",
    "                          'T_DN(Q)' : [2,3,4,\n",
    "                                       12, 20, 31, 10, 29,  9, 18, 26, 23, 32, 21, 16, 15,  7, 13, 27, 14, 30, 22, 19],\n",
    "                          'T_DN(late)' : [0],\n",
    "                          'Remove' : [25,35]} # LowQC and doublets\n",
    "\n",
    "np.array([c for c in adata.obs['leiden_r2.0'].unique().tolist() if c not in [c for c in dn_cluster_assignments.values() for c in c]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['temp_anno'] = pd.NA\n",
    "#dn_cluster_assignments = {k: [c for c in dn_cluster_assignments[k]] for k in dn_cluster_assignments.keys()}\n",
    "for k, v in dn_cluster_assignments.items():\n",
    "    adata.obs.loc[adata.obs['leiden_r2.5'].isin(v), 'temp_anno'] = k\n",
    "    \n",
    "sc.pl.umap(adata, color = 'temp_anno', return_fig = True)\n",
    "plt.savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_tempAnno_umap.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata, color = ['CD34','CD3D', 'VPREB1', 'DNTT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[adata.obs['temp_anno'] == 'B_dev_thy']['donor'].value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata, \n",
    "                ct_markers_filtered,\n",
    "                categories_order = ['B_dev_thy', 'T_DN(early)', 'T_DN(P)', 'T_DN(Q)', 'T_DN(late)', 'Remove'],\n",
    "                groupby = 'temp_anno',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_dnMarkers_tempAnno_dotplot.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['temp_anno'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs[['temp_anno']].to_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_dnSplit_scvi_{object_version}_tempAnno.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check cell cycle activity:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Score cell cycle\n",
    "cell_cycle_genes_df = pd.read_excel('/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/T_NK_compartment/data/curated/regev_cell_cycle_genes.xlsx', skiprows = 6, usecols=[0,1])\n",
    "\n",
    "cell_cycle_genes = {}\n",
    "for col in cell_cycle_genes_df.columns:\n",
    "    cell_cycle_genes_df[col] = cell_cycle_genes_df[col].str.strip()\n",
    "    cell_cycle_genes[col] = [g.strip() for g in cell_cycle_genes_df[col].dropna().tolist() if g.strip() in adata.var_names]\n",
    "    \n",
    "import pprint\n",
    "pprint.pprint(cell_cycle_genes, compact=True)\n",
    "\n",
    "sc.tl.score_genes_cell_cycle(adata, s_genes=cell_cycle_genes['G1/S'], g2m_genes=cell_cycle_genes['G2/M'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_cycle_score_df = adata.obs[['S_score', 'G2M_score', 'donor', 'age_group', 'temp_anno', 'phase', 'study']].copy()\n",
    "#cell_cycle_score_df = cell_cycle_score_df.join(knn_anno)\n",
    "df = cell_cycle_score_df.groupby(['age_group', 'temp_anno', 'donor', 'study'], observed = True)['phase'].value_counts(normalize = True).unstack().fillna(0).reset_index()\n",
    "df = df.melt(id_vars = ['age_group', 'temp_anno', 'donor', 'study'], value_name = 'prop', var_name = 'phase')\n",
    "df['temp_anno'] = pd.Categorical(df['temp_anno'], categories = ['T_ETP','T_DN(early)', 'T_DN(P)', 'T_DN(Q)', 'T_DN(late)', 'B_dev_thy'], ordered=True)\n",
    "df['temp_anno'] = df['temp_anno'].cat.remove_unused_categories()\n",
    "\n",
    "sns.catplot(data=df, x='temp_anno', y='prop', hue = 'age_group', hue_order = ['infant', 'paed', 'adult', 'geriatric'], kind=\"box\", palette = 'colorblind',\n",
    "            col='phase', col_order=['G1', 'S','G2M'], \n",
    "            #facet_kws={'sharey': False, 'sharex': True}, \n",
    "            height=2, aspect=2, col_wrap=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CD4 recirc T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata_cd4_recirc\n",
    "object_version = 'v4_2025-02-04'\n",
    "adata_cd4_recirc = ad.read_h5ad(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_{object_version}.zarr', backed = 'r')\n",
    "\n",
    "# Add latest annotations to adata_cd4_recirc\n",
    "ct_anno = pd.read_csv(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_{object_version}_curatedAnno_v8.csv', index_col = 0)\n",
    "for c in ct_anno.columns:\n",
    "    if c in adata_cd4_recirc.obs.columns:\n",
    "        adata_cd4_recirc.obs.drop(c, axis = 1, inplace = True)\n",
    "adata_cd4_recirc.obs = adata_cd4_recirc.obs.join(ct_anno)\n",
    "\n",
    "adata_cd4_recirc = adata_cd4_recirc[adata_cd4_recirc.obs['taa_l3'].isin(['T_CD4_recirc', 'T_Treg_recirc'])].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scvi_wrapper import run_scvi\n",
    "object_version = f'v1_{today}'\n",
    "\n",
    "# Run scvi\n",
    "scvi_run = run_scvi(adata_cd4_recirc, \n",
    "                    layer_raw = 'X', \n",
    "                    # Excluded genes\n",
    "                    include_genes=[], exclude_cc_genes=True, exclude_mt_genes=True, \n",
    "                    exclude_vdjgenes = True, remove_cite = False,\n",
    "                    # Highly variable gene selection\n",
    "                    batch_hv=\"study\", hvg = 200, #span = 1,\n",
    "                    hvg_selection = 'experimental',\n",
    "                    # scVI \n",
    "                    batch_scvi=\"sample\",\n",
    "                    cat_cov_scvi=[\"donor\", \"chemistry_simple\", \"sex\"], \n",
    "                    cont_cov_scvi=[], \n",
    "                    max_epochs=400, batch_size=2000, early_stopping = True, early_stopping_patience = 15, early_stopping_min_delta = 10.0,\n",
    "                    plan_kwargs = {'lr': 0.001, 'reduce_lr_on_plateau' : True, 'lr_patience' : 10, 'lr_threshold' : 20}, \n",
    "                    n_layers = 3, n_latent = 30, dispersion = 'gene-batch',\n",
    "                    # Leiden clustering\n",
    "                    leiden_clustering = None, col_cell_type = ['taa_l5'], \n",
    "                    fig_dir = f'{plots_path}/ctAnnotation/v9', fig_prefix = f'thyAgeing_cd4RecircSplit_scvi_{object_version}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save adata and scvi model\n",
    "overwrite = True\n",
    "\n",
    "anno_cols = [c for c in scvi_run['data'].obs.columns if '_pred_' in c or '_prob_' in c or 'taa' in c or 'anno_' in c or 'leiden' in c]\n",
    "if not os.path.exists(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}.zarr') or overwrite:\n",
    "    scvi_run['data'].obs = scvi_run['data'].obs.drop(columns=anno_cols)\n",
    "    scvi_run['data'].write_h5ad(\n",
    "        f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_v2.zarr',\n",
    "        compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "        compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,\n",
    "    )\n",
    "    scvi_run['vae'].save(f'{model_path}/thyAgeing_cd4RecircSplit_scvi_{object_version}', save_anndata=False, overwrite=overwrite)\n",
    "else:\n",
    "    print('File already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Leiden clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Louvain clustering\n",
    "adata_cd4_recirc = ad.read_h5ad(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_v2.zarr')\n",
    "\n",
    "res_list = [1.0, 1.5]\n",
    "for res in res_list:\n",
    "    sc.tl.leiden(adata_cd4_recirc, resolution = res, key_added = f\"leiden_r{res}\")\n",
    "adata_cd4_recirc.obs[[f'leiden_r{str(r)}' for r in res_list]] = adata_cd4_recirc.obs[[f'leiden_r{str(r)}' for r in res_list]].astype(int).astype('category')\n",
    "\n",
    "adata_cd4_recirc.obs[[f'leiden_r{str(r)}' for r in res_list]].to_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_leidenClusters.csv')\n",
    "\n",
    "sc.pl.umap(adata_cd4_recirc, color = [f'leiden_r{str(r)}' for r in res_list], show = False, return_fig = True)\n",
    "plt.savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_leidenClusters.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Marker expression and ct annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "object_version = 'v1_2025-03-28'\n",
    "adata_cd4_recirc = ad.read_h5ad(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_v2.zarr')\n",
    "leiden_clustering = pd.read_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_leidenClusters.csv', index_col = 0)\n",
    "adata_cd4_recirc.obs = adata_cd4_recirc.obs.join(leiden_clustering)\n",
    "adata_cd4_recirc.obs[leiden_clustering.columns] = adata_cd4_recirc.obs[leiden_clustering.columns].astype(int).astype('category')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.normalize_total(adata_cd4_recirc, target_sum=1e4)\n",
    "sc.pp.log1p(adata_cd4_recirc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load curated markers and make dictionary\n",
    "ct_markers_df = pd.read_excel(f'{data_path}/curated/curatedCellMarkers_LMM.xlsx', sheet_name = 'Final')\n",
    "ct_markers = ct_markers_df.groupby('cell_label_fine')['marker_gene'].agg(list).to_dict()\n",
    "\n",
    "cd4_markers = {'T_CD4_recirc' : ['PRDM1', 'PTPRC', 'LAG3', 'HAVCR2', 'CTLA4', 'PDCD1', 'CCR7', 'S1PR1'],\n",
    "               'T_CD4_r1_recirc':['LAG3', 'IL10', 'TGFB1', 'PRDM1', 'GZMA', 'IFNG','PDCD1'],\n",
    "               'T_T(agonist)':['NR4A1', 'NR4A2', 'NR4A3', 'BCL2L11'],\n",
    "               'T_CD4_FH' : ['CD40LG', 'PDCD1', 'BCL6', 'IL21', 'CXCL13', 'SLAMF6', 'CD200', 'MAF', 'CXCR5'],\n",
    "               'T_CD4_h1' : ['IFNG', 'TNF', 'TBX21', 'STAT1', 'STAT4', 'CXCR3', 'IL12RB2', 'LY6E', 'GZMK'],\n",
    "               'T_CD4_h17' : ['IL17A', 'IL17F', 'IL22', 'IL21', 'RORC', 'STAT3', 'AHR', 'CCR6', 'IL23R', 'CCL20'],\n",
    "}\n",
    "\n",
    "# Filter markers of interest\n",
    "ctoi = ['T_CD4','T_CD4_fh','T_CD4_h1','T_CD4_h17','T_CD4_r1','T_CD4_recirc', 'T_Treg']\n",
    "ct_markers_filtered = {c: list(set([g for g in ct_markers[c] if g in adata_cd4_recirc.var_names])) for c in ctoi if c in ct_markers}\n",
    "ct_markers_filtered['TCR_act'] = ['NR4A1', 'NR4A2', 'NR4A3']\n",
    "ct_markers_filtered['T_CD4_fr'] =  ['FOXP3', 'IL2RA', 'CTLA4', 'IL10', 'TGFB1']\n",
    "ct_markers_filtered['T_CD4_mem'] = ['FOSB', 'KLF6', 'BCL2', 'SELL', 'IFNG', 'FOXP1', 'LEF1', 'IL7R']\n",
    "ct_markers_filtered['T_CD4_h2'] = ['GATA3', 'STAT6', 'BATF', 'IRF4', 'IL4', 'IL5', 'IL13', 'CCR4', 'CCR8', 'PTGDR2']\n",
    "ct_markers_filtered['Homing'] = ['S1PR1', 'CCR7', 'CXCR6', 'CXCR3', 'CCR4', 'CCR5', 'CCR6', 'CCR8']\n",
    "ct_markers_filtered['TNF signalling'] = ['TNFRSF18', 'TNFRSF4', 'TNFRSF1B']\n",
    "ct_markers_filtered['Activation'] = ['IL2RA', 'TIGIT', 'ICOS', 'LMNA']\n",
    "ct_markers_filtered['Proliferation'] = ['TOP2A', 'MKI67']\n",
    "ct_markers_filtered['T_CD4_fh'].extend(['SH2D1A', 'BATF'])\n",
    "\n",
    "plt.rcParams.update(plt.rcParamsDefault)\n",
    "sc.pl.DotPlot(adata_cd4_recirc, \n",
    "                ct_markers_filtered,\n",
    "                groupby = 'leiden_r1.5',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_cd4Markers_leiden1.5_dotplot.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[c for c in ct_markers_filtered['T_CD4_h2'] if not c in adata_cd4_recirc.var_names]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_cluster_assignments = {'T_CD4_naive_recirc' : [4,10,7],\n",
    "                          'T_CD4_fh' : [3],\n",
    "                          'T_CD4_unstim' : [1,5,8,15,21,23,9,24],\n",
    "                          'T_CD4_h' : [11,13,19,20], # CM CCR7+, EM CCR6+/CXCR3+\n",
    "                          'T_CD4_act' : [6], # CCR4\n",
    "                          'T_Treg_fr' : [],\n",
    "                          'T_Treg_recirc' : [2,12,17,14,18],\n",
    "                          'T_Treg_tr' : [0,16,22],\n",
    "                          'Remove' : [25]}\n",
    "\n",
    "np.array([c for c in adata_cd4_recirc.obs['leiden_r1.5'].unique().tolist() if c not in [c for c in cd4_cluster_assignments.values() for c in c]])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Single-cell RNA sequencing (scRNA-seq) allows for the identification of specific gene expression profiles that can serve as markers for different cell types, including CD4 memory T cells. Here are some key markers typically associated with CD4 memory T cells:\n",
    "\n",
    "1. **CD45RO**: A classic marker distinguishing memory T cells from naive T cells (which express CD45RA).\n",
    "2. **CCR7**: Central memory T cells (T_CM) express CCR7, which is involved in homing to lymphoid tissues.\n",
    "3. **CD27**: Often expressed on memory T cells, particularly central memory T cells.\n",
    "4. **CD62L (L-selectin)**: Another marker for central memory T cells, involved in lymph node homing.\n",
    "5. **IL7R (CD127)**: Expressed on memory T cells, indicating their responsiveness to IL-7, which is important for their survival.\n",
    "6. **CXCR3**: Often expressed on effector memory T cells (T_EM), involved in migration to inflamed tissues.\n",
    "7. **CCR6**: Expressed on a subset of memory T cells, particularly Th17 cells.\n",
    "8. **CD69**: An early activation marker that can be transiently expressed on memory T cells.\n",
    "9. **CD95 (Fas)**: Expressed on memory T cells, involved in apoptosis regulation.\n",
    "10. **BCL2**: Anti-apoptotic protein, often upregulated in memory T cells to promote their survival.\n",
    "11. **KLRG1**: Expressed on a subset of effector memory T cells, often associated with terminal differentiation.\n",
    "12. **HLA-DR**: MHC class II molecule, sometimes expressed on activated memory T cells.\n",
    "13. **IFNG (Interferon-gamma)**: Cytokine often produced by memory T cells, particularly Th1 cells.\n",
    "14. **GZMB (Granzyme B)**: Expressed by cytotoxic memory T cells, involved in target cell killing.\n",
    "15. **PRF1 (Perforin)**: Another marker of cytotoxic memory T cells, involved in target cell lysis.\n",
    "\n",
    "These markers can be used to identify and characterize CD4 memory T cells in single-cell RNA sequencing data, providing insights into their functional states and roles in immune responses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_cluster_assignments = {'T_CD4_naive_recirc' : [24,  1,  9,  8,  4, 10, 23, 21,  7,  5, 15],\n",
    "                          'T_CD4_fh' : [3],\n",
    "                          'T_CD4_h' : [20,11,13,19,20],\n",
    "                          'T_CD4_act' : [6],\n",
    "                          'T_Treg_fr' : [],\n",
    "                          'T_Treg_HSPhi' : [2,12,17,14],\n",
    "                          'T_Treg_recirc' : [0,16,22],\n",
    "                          'Remove' : [25]}\n",
    "\n",
    "np.array([c for c in adata_cd4_recirc.obs['leiden_r1.5'].unique().tolist() if c not in [c for c in cd4_cluster_assignments.values() for c in c]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cd4_recirc.obs['temp_anno'] = pd.NA\n",
    "#dn_cluster_assignments = {k: [c for c in dn_cluster_assignments[k]] for k in dn_cluster_assignments.keys()}\n",
    "for k, v in cd4_cluster_assignments.items():\n",
    "    adata_cd4_recirc.obs.loc[adata_cd4_recirc.obs['leiden_r1.5'].isin(v), 'temp_anno'] = k\n",
    "    \n",
    "sc.pl.umap(adata_cd4_recirc, color = 'temp_anno', return_fig = True)\n",
    "plt.savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_tempAnno_umap.pdf', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(adata_cd4_recirc, groupby = 'temp_anno', n_genes = 30, method = 'wilcoxon', reference='rest')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.get.rank_genes_groups_df(adata_cd4_recirc, group = ['T_CD4_act'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.DotPlot(adata_cd4_recirc, \n",
    "                ct_markers_filtered,\n",
    "                groupby = 'temp_anno',\n",
    "                mean_only_expressed=True,\n",
    "                cmap = 'magma').add_totals().savefig(f'{plots_path}/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_cd4Markers_tempAnno_dotplot.png', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "HSPhi T cell: https://www.sciencedirect.com/science/article/pii/S2589004223016656"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_cd4_recirc.obs[['temp_anno']].to_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_{object_version}_tempAnno.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Whole T/NK object \n",
    "\n",
    "### Integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata_full\n",
    "object_version = 'v4_2025-02-04'\n",
    "adata_full = ad.read_h5ad(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_{object_version}.zarr')\n",
    "\n",
    "# Add latest annotations to adata_full\n",
    "ct_anno = pd.read_csv(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_{object_version}_curatedAnno_v8.csv', index_col = 0)\n",
    "for c in ct_anno.columns:\n",
    "    if c in adata_full.obs.columns:\n",
    "        adata_full.obs.drop(c, axis = 1, inplace = True)\n",
    "adata_full.obs = adata_full.obs.join(ct_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes = ct_anno.index[ct_anno['taa_l1'].isin(['T', 'NK'])].to_list()\n",
    "print(len(barcodes))\n",
    "barcodes = set(barcodes + adata.obs_names.tolist())\n",
    "\n",
    "len(barcodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full = adata_full.concatenate(adata_c9, index_unique = None).copy()\n",
    "adata_full = adata_full[adata_full.obs_names.isin(barcodes)].to_memory()\n",
    "\n",
    "adata_full.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "object_version = f'v9_{today}'\n",
    "\n",
    "# Run scvi\n",
    "scvi_run = run_scvi(adata_full, \n",
    "                    layer_raw = 'X', \n",
    "                    # Excluded genes\n",
    "                    include_genes=[], exclude_cc_genes=True, exclude_mt_genes=True, \n",
    "                    exclude_vdjgenes = True, remove_cite = False,\n",
    "                    # Highly variable gene selection\n",
    "                    batch_hv=\"study\", hvg = 3000, \n",
    "                    hvg_selection = 'experimental',\n",
    "                    # scVI \n",
    "                    batch_scvi=\"sample\",\n",
    "                    cat_cov_scvi=[\"donor\", \"chemistry_simple\", \"sex\"], \n",
    "                    #cont_cov_scvi=[\"percent_mito\", 'percent_ribo', 'n_genes'], \n",
    "                    max_epochs=400, batch_size=2000, early_stopping = True, early_stopping_patience = 15, early_stopping_min_delta = 10.0,\n",
    "                    plan_kwargs = {'lr': 0.001, 'reduce_lr_on_plateau' : True, 'lr_patience' : 10, 'lr_threshold' : 20}, \n",
    "                    n_layers = 3, n_latent = 30, dispersion = 'gene-batch',\n",
    "                    # Leiden clustering\n",
    "                    leiden_clustering = None, col_cell_type = ['taa_l5'], \n",
    "                    fig_dir = f'{plots_path}/preprocessing', fig_prefix = f'thyAgeing_tSplit_scvi_{object_version}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save adata and scvi model\n",
    "overwrite = True\n",
    "\n",
    "anno_cols = [c for c in scvi_run['data'].obs.columns if '_pred_' in c or '_prob_' in c or 'taa' in c or 'anno_' in c or 'leiden' in c]\n",
    "if not os.path.exists(f'{data_path}/thyAgeing_tSplit_scvi_{object_version}.zarr') or overwrite:\n",
    "    scvi_run['data'].obs = scvi_run['data'].obs.drop(columns=anno_cols)\n",
    "    scvi_run['data'].write_h5ad(\n",
    "        f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_{object_version}.zarr',\n",
    "        compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "        compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,\n",
    "    )\n",
    "    scvi_run['vae'].save(f'{model_path}/thyAgeing_tSplit_scvi_{object_version}', save_anndata=False, overwrite=overwrite)\n",
    "else:\n",
    "    print('File already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "object_version = 'v9_2025-03-28'\n",
    "adata_full = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_{object_version}.zarr')\n",
    "\n",
    "# Add latest annotations to adata\n",
    "ct_anno = pd.read_csv(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_v4_2025-02-04_curatedAnno_v8.csv', index_col = 0)\n",
    "for c in ct_anno.columns:\n",
    "    if c in adata_full.obs.columns:\n",
    "        adata_full.obs.drop(c, axis = 1, inplace = True)\n",
    "adata_full.obs = adata_full.obs.join(ct_anno)\n",
    "\n",
    "# Update metadata\n",
    "latest_meta_path = get_latest_version(dir = f'{general_data_path}/metadata', file_prefix='Thymus_ageing_metadata')\n",
    "latest_meta = pd.read_excel(latest_meta_path)\n",
    "update_obs(adata_full, latest_meta, on = 'index', ignore_warning = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full.obs['temp_anno'] = adata_full.obs['taa_l5'].astype(str)\n",
    "\n",
    "dn_anno = pd.read_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_dnSplit_all_scvi_v4_2025-02-04_tempAnno.csv', index_col = 0)\n",
    "cd4_recirc_anno = pd.read_csv(f'{data_path}/preprocessing/ctAnnotation/v9/thyAgeing_cd4RecircSplit_scvi_v1_2025-03-28_tempAnno.csv', index_col = 0)\n",
    "\n",
    "adata_full.obs.loc[dn_anno.index,'temp_anno'] = dn_anno['taa_l5'].values\n",
    "adata_full.obs.loc[cd4_recirc_anno.index,'temp_anno'] = cd4_recirc_anno['temp_anno'].values\n",
    "adata_full.obs.loc[adata_full.obs['temp_anno'] == 'T_CD4_prolif','temp_anno'] = 'T_CD4_act'\n",
    "\n",
    "adata_full.obs['temp_anno'] = adata_full.obs['temp_anno'].astype('category')\n",
    "adata_full.obs.loc[adata_full.obs['temp_anno'].isin(['Remove', 'nan']),'temp_anno'] = pd.NA\n",
    "adata_full.obs['temp_anno'].cat.remove_unused_categories()\n",
    "\n",
    "adata_full.obs['temp_anno'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full.obs_names.duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.umap(adata_full[~adata_full.obs['temp_anno'].isna()], color = 'temp_anno')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full.obs[['temp_anno']].to_csv(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_{object_version}_tempAnno.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_levels = pd.read_excel(f'{general_data_path}/curated/thyAgeing_full_curatedAnno_v9_2025-03-03_levels.xlsx')\n",
    "ct_anno = adata_full.obs[['temp_anno']].copy()\n",
    "ct_anno = ct_anno.reset_index(names='barcodes').merge(anno_levels, left_on = 'temp_anno', right_on = 'taa_l5', how = 'inner').drop(columns = ['temp_anno']).set_index('barcodes')\n",
    "ct_anno.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_anno['taa_l5'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check number of NAs (removed cells)\n",
    "ct_anno.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check whether all cell types are included in reference\n",
    "np.setdiff1d(adata_full.obs['temp_anno'].unique().astype(str), ct_anno['taa_l5'].unique().astype(str))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_anno.to_csv(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_{object_version}_curatedAnno_v9.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotting.utils import plot_grouped_boxplot, calc_figsize\n",
    "from anno_levels import get_ct_levels, get_ct_palette, age_group_levels, age_group_palette\n",
    "\n",
    "col_cell_type_fine = 'taa_l5'\n",
    "col_age_group = 'age_group'\n",
    "col_age_group_levels = age_group_levels\n",
    "col_cell_type_fine_levels = [c for c in get_ct_levels('taa_l5') if c in adata_full.obs[col_cell_type_fine].unique().tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_full.obs.drop(columns = ct_anno.columns, inplace = True, errors='ignore')\n",
    "adata_full.obs = adata_full.obs.join(ct_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno_df.groupby(['taa_l5', col_age_group, 'donor']).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create anno df\n",
    "anno_df = adata_full.obs[['sample', 'donor', 'sex', 'sort', 'study', col_age_group, 'age_months', col_cell_type_fine,]].copy()\n",
    "anno_df.dropna(subset = col_cell_type_fine, inplace = True)\n",
    "anno_df_sub = anno_df[(anno_df['sort'].isin(['TOT', 'CD3P']))].copy()\n",
    "df = freq_by_donor(anno_df_sub, sample_col = 'sample', donor_col = 'donor', summary_col=col_cell_type_fine, add_meta = [col_age_group])\n",
    "\n",
    "plot_grouped_boxplot(data = df, x = col_cell_type_fine, y = 'mean_prop', hue = col_age_group, order = col_cell_type_fine_levels, hue_order = col_age_group_levels, \n",
    "                     palette = age_group_palette, x_label = 'Cell population', y_label = 'Frequency', legend_title = 'Age group', add_stats = True, format_percent = True, figsize = calc_figsize(height = 'half', height_ratio = 0.75),\n",
    "                     #save_stats = f'{data_path}/analyses/freqAnalysis/all_curatedAnno_v7/thyAgeing_tSplit_relTfreqFine'\n",
    "                     )\n",
    "plt.savefig(f'{plots_path}/ctAnnotation/v9/testFreq.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate transition rates\n",
    "transitions = [('T_DN(early)', 'T_DN(P)'), ('T_DN(P)', 'T_DN(Q)'), ('T_DN(Q)', 'T_DP(P)'), ('T_DP(P)', 'T_DP(Q)'), ('T_DP(Q)', 'T_αβT(entry)'), \n",
    "               ('T_αβT(entry)', 'T_CD4_naive'), ('T_αβT(entry)', 'T_CD8_naive'), ('T_αβT(entry)', 'T_Treg')]\n",
    "transitions_order = [f'{t0} -> {t1}' for t0, t1 in transitions]\n",
    "\n",
    "trans_rates = []\n",
    "for trans in transitions:\n",
    "    df = anno_df[(anno_df[col_cell_type_fine].isin(trans)) & (anno_df['sort'].isin(['TOT', 'CD3P']))]\n",
    "    \n",
    "    pairwise_combinations = pd.MultiIndex.from_product([df['sample'].unique(), df[col_cell_type_fine].unique()], names=['sample', col_cell_type_fine]).to_frame(index=False)\n",
    "    pairwise_combinations = pairwise_combinations.merge(df[['sample','donor', col_age_group]], on = 'sample', how = 'left')\n",
    "    \n",
    "    df = df.groupby([col_age_group, col_cell_type_fine, 'sample', 'donor'], observed = True).size().reset_index(name='counts')\n",
    "    df = df.merge(pairwise_combinations, on = [col_cell_type_fine, col_age_group, 'sample', 'donor'], how = 'outer').fillna(0)\n",
    "    df = df.pivot_table(index=[col_age_group, 'sample', 'donor'], columns=col_cell_type_fine, values='counts', aggfunc='sum').fillna(0).reset_index()\n",
    "\n",
    "    df['ratio'] = np.where(df[trans[0]] == 0, np.nan, (df[trans[1]] / df[trans[0]]) + 0.001)\n",
    "    df['transition'] = f'{trans[0]} -> {trans[1]}'\n",
    "    \n",
    "    df = df.groupby([col_age_group, 'transition', 'donor'], observed=True).agg(ratio = ('ratio', 'mean')).reset_index()\n",
    "    \n",
    "    trans_rates.append(df)\n",
    "    \n",
    "trans_rates = pd.concat(trans_rates)\n",
    "\n",
    "trans_rates.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Formatting\n",
    "from matplotlib import font_manager\n",
    "font_manager.fontManager.addfont(\"/nfs/team205/ny1/ThymusSpatialAtlas/software/Arial.ttf\")\n",
    "plt.style.use('/lustre/scratch126/cellgen/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts/plotting/thyAgeing.mplstyle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_grouped_boxplot(data = trans_rates.reset_index(), x = 'transition', y = 'ratio', hue = col_age_group, order = transitions_order, hue_order = col_age_group_levels, \n",
    "                     x_label = 'Transition', y_label = 'Transition rate', legend_title = 'Age group', add_stats = True, format_log = True, figsize = calc_figsize(height = 70, width = 73), \n",
    "                     y_intercept = 1.0, ylim = (0.1, 5000),\n",
    "                     #save_stats = f'{data_path}/analyses/freqAnalysis/all_curatedAnno_v7/thyAgeing_tSplit_fateTransitions'\n",
    "                     )\n",
    "plt.savefig(f'{plots_path}/ctAnnotation/v9/testTrans.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nhood_enrichment = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_tSplitxTissue_scvi_v1_2024-11-29_nhoodEnrichment.csv', index_col=0)\n",
    "nhood_enrichment.replace(pd.NA, 'none', inplace = True)\n",
    "nhood_enrichment = nhood_enrichment.loc[nhood_enrichment.index.str.endswith('-0')]\n",
    "nhood_enrichment.index = nhood_enrichment.index.str.replace('-0', '')\n",
    "anno_df = anno_df.join(nhood_enrichment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nhood_enrichment['nhood_enrichment'].value_counts() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_location = anno_df[(anno_df['sort'].isin(['TOT', 'CD3P'])) & (anno_df[col_cell_type_fine].isin(col_cell_type_fine_levels))].groupby(['donor', col_cell_type_fine])['nhood_enrichment'].value_counts(normalize=True).to_frame('freq').reset_index()\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# Convert 'freq' to percentage\n",
    "df_location['freq'] = df_location['freq']\n",
    "\n",
    "# Plot grouped barplot\n",
    "plt.figure(figsize=calc_figsize(width_ratio = 1, height_ratio = 0.5))\n",
    "ax = plt.axes()\n",
    "sns.barplot(data=df_location, x=col_cell_type_fine, y='freq', hue='nhood_enrichment', order = [l for l in col_cell_type_fine_levels if l in df_location[col_cell_type_fine].unique()], hue_order = ['blood', 'tissue', 'none'], ci='sd')\n",
    "plt.xlabel('Cell population')\n",
    "plt.ylabel('Proportion')\n",
    "ax.yaxis.set_major_formatter(plt.FuncFormatter(lambda x, _: f'{x:.0%}'))\n",
    "plt.xticks(rotation=90)\n",
    "plt.legend(title='Location enrichment')\n",
    "plt.tight_layout()\n",
    "#plt.savefig(f'{plots_path}/freqAnalysis/all_curatedAnno_v7/thyAgeing_tSplit_cellType_by_nhood.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blood_enriched = anno_df[(anno_df['sort'].isin(['TOT', 'CD3P'])) & (anno_df['nhood_enrichment'] == 'blood')].groupby(['donor', col_age_group])[col_cell_type_fine].value_counts(normalize=True).to_frame('freq').reset_index().dropna()\n",
    "tissue_enriched = anno_df[(anno_df['sort'].isin(['TOT', 'CD3P'])) & (anno_df['nhood_enrichment'] == 'tissue')].groupby(['donor', col_age_group])[col_cell_type_fine].value_counts(normalize=True).to_frame('freq').reset_index().dropna()\n",
    "df = tissue_enriched.merge(blood_enriched, on = ['donor', col_age_group, col_cell_type_fine], suffixes = ('_tissue', '_blood'))\n",
    "df['ratio'] = df['freq_tissue'] / df['freq_blood']\n",
    "df.dropna(inplace = True)  \n",
    "df = df.loc[~df[col_cell_type_fine].isin(['T_CD8αα(entry)', # Exclude innate\n",
    " 'T_CD8αα(I)',\n",
    " 'T_CD8αα(II)',\n",
    " 'T_MAIT',\n",
    " 'ILC',\n",
    " 'T_γδT',\n",
    " 'NK_tr'])] \n",
    "#df[col_cell_type_fine].cat.remove_unused_categories(inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_grouped_boxplot(data = df, x = col_cell_type_fine, y = 'ratio', hue = col_age_group, order = [c for c in col_cell_type_fine_levels if c in df[col_cell_type_fine].unique()], hue_order = col_age_group_levels, \n",
    "                     x_label = 'Cell population', y_label = 'Ratio [tissue/blood]', legend_title = 'Age group', add_stats = True, format_log = True, figsize = calc_figsize(width = 300, height = 300), ylim = (0.01,100), y_intercept = 1,\n",
    "                     #save_stats = f'{data_path}/analyses/freqAnalysis/all_curatedAnno_v7/thyAgeing_tSplit_ratioTissueBlood'\n",
    ")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
