{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Trajectory analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import session_info\n",
    "from datetime import datetime\n",
    "today = datetime.today().strftime('%Y-%m-%d')\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import hdf5plugin\n",
    "import scFates as scf\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\", category=ad.ImplicitModificationWarning)\n",
    "\n",
    "# Add repo path to sys path (allows to access scripts and metadata from repo)\n",
    "#repo_path,_ = os.path.split(os.path.split(os.getcwd())[0])\n",
    "repo_path = '/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/T_NK_compartment'\n",
    "sys.path.insert(1, repo_path) \n",
    "sys.path.insert(2, '/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts')\n",
    "\n",
    "# Autoreload custom scripts\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "# Define paths\n",
    "plots_path = f'{repo_path}/plots/'\n",
    "data_path = f'{repo_path}/data/'\n",
    "model_path = os.path.join(repo_path, 'models')\n",
    "general_data_path = '/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/data'\n",
    "\n",
    "print('Dir for plots: {}'.format(plots_path))\n",
    "print('Dir for data: {}'.format(data_path))\n",
    "\n",
    "# Formatting\n",
    "from matplotlib import font_manager\n",
    "sc.settings.set_figure_params(dpi = 150, color_map = 'RdPu', dpi_save = 300, vector_friendly = True, format = 'pdf')\n",
    "font_manager.fontManager.addfont(\"/nfs/team205/ny1/ThymusSpatialAtlas/software/Arial.ttf\")\n",
    "plt.style.use('/nfs/team205/lm25/thymus_projects/thymus_ageing_atlas/General_analysis/scripts/plotting/thyAgeing.mplstyle')\n",
    "\n",
    "# Import custom scripts\n",
    "from utils import get_latest_version,update_obs,freq_by_donor\n",
    "from anno_levels import get_ct_levels, get_ct_palette, age_group_levels, age_group_palette, t_nk_groupings\n",
    "from plotting.utils import plot_grouped_boxplot, calc_figsize,thyAgeing_colors,thyAgeing_greys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata\n",
    "object_version = 'v5_2025-04-03'\n",
    "adata = ad.read_h5ad(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_{object_version}.zarr')\n",
    "\n",
    "# Add new annotations to adata\n",
    "ct_anno = pd.read_csv(f'{general_data_path}/objects/rna/thyAgeing_all_scvi_v4_2025-02-04_curatedAnno_v10.csv', index_col = 0)\n",
    "adata.obs = adata.obs.join(ct_anno, how = 'left')\n",
    "\n",
    "# Update metadata\n",
    "latest_meta_path = get_latest_version(dir = f'{general_data_path}/metadata', file_prefix='Thymus_ageing_metadata')\n",
    "latest_meta = pd.read_excel(latest_meta_path)\n",
    "update_obs(adata, latest_meta, on = 'index', ignore_warning = True)\n",
    "\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['taa_l4'].value_counts()['NK_tr']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DN(early) bottleneck"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define which cell states to select\n",
    "ctoi = ['B_dev_thy','T_DN(early)', 'T_DN(P)', 'pDC', 'ILC', 'NK_tr', 'T_MAIT'] # 'CMP', 'Myelocyte', 'Neutrophil', 'Promonocyte', 'Mono', 'DC1', 'DC1-prolif',\n",
    "        \n",
    "\n",
    "# Define columns\n",
    "col_cell_type_fine = 'taa_l4'\n",
    "col_cell_type_fine_levels = get_ct_levels(col_cell_type_fine, include_ct=ctoi)\n",
    "col_age_group = 'age_group'\n",
    "col_age_group_levels = eval(f'{col_age_group}_levels')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_cell_type_fine_levels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset adata\n",
    "adata_sub = adata[adata.obs[col_cell_type_fine].isin(ctoi)]\n",
    "\n",
    "# # Update metadata\n",
    "# latest_meta_path = get_latest_version(dir = f'{general_data_path}/metadata', file_prefix='Thymus_ageing_metadata')\n",
    "# latest_meta = pd.read_excel(latest_meta_path)\n",
    "# update_obs(adata_sub, latest_meta, on = 'index', ignore_warning = True)\n",
    "\n",
    "adata_sub.obs[col_cell_type_fine].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate PCA & UMAP embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove cell cycle genes\n",
    "cell_cycle_genes = [x.strip() for x in open('/nfs/team205/vk8/processed_data/regev_lab_cell_cycle_genes.txt')]\n",
    "adata_sub = adata_sub[:,~adata_sub.var.index.isin(cell_cycle_genes)]\n",
    "\n",
    "# Log-normalize\n",
    "adata_sub.layers['counts'] = adata_sub.X.copy()\n",
    "sc.pp.normalize_total(adata_sub, target_sum=1e4)\n",
    "sc.pp.log1p(adata_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub.obs['study'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove studies with fewer than 50 cells\n",
    "min_cells = 10\n",
    "study_counts = adata_sub.obs['study'].value_counts()\n",
    "studies_to_keep = study_counts[study_counts >= min_cells].index\n",
    "adata_sub = adata_sub[adata_sub.obs['study'].isin(studies_to_keep)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select HVG and run PCA\n",
    "sc.pp.highly_variable_genes(adata_sub, n_top_genes=500, flavor='seurat', n_bins=20, batch_key = 'study')\n",
    "sc.pp.pca(adata_sub, n_comps=10, use_highly_variable=True, svd_solver='arpack')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.pca_variance_ratio(adata_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Perform UMAP\n",
    "sc.pp.neighbors(adata_sub, n_pcs=8, n_neighbors=20, random_state=42)\n",
    "sc.tl.umap(adata_sub)\n",
    "\n",
    "with plt.rc_context({'figure.figsize' : calc_figsize(width = 70, height = 40)}):\n",
    "    sc.pl.umap(adata_sub, color=[col_cell_type_fine, col_age_group], \n",
    "               ncols=2, wspace=0.3, return_fig=True, show = False, size = 1, frameon = False)\n",
    "    plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_umap.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scvi_wrapper import run_scvi\n",
    "object_version = f'v1_{today}'\n",
    "\n",
    "# Run scvi\n",
    "scvi_run = run_scvi(adata_sub, \n",
    "                    layer_raw = 'X', \n",
    "                    # Excluded genes\n",
    "                    include_genes=[], exclude_cc_genes=True, exclude_mt_genes=True, \n",
    "                    exclude_vdjgenes = True, remove_cite = False,\n",
    "                    # Highly variable gene selection\n",
    "                    batch_hv=\"study\", hvg = 1000, span = 0.5,\n",
    "                    hvg_selection = 'experimental',\n",
    "                    # scVI \n",
    "                    batch_scvi=\"sample\",\n",
    "                    cat_cov_scvi=[\"donor\", \"chemistry_simple\", \"sex\"], \n",
    "                    cont_cov_scvi=[], \n",
    "                    max_epochs=400, batch_size=2000, early_stopping = True, early_stopping_patience = 15, early_stopping_min_delta = 10.0,\n",
    "                    plan_kwargs = {'lr': 0.001, 'reduce_lr_on_plateau' : True, 'lr_patience' : 10, 'lr_threshold' : 20}, \n",
    "                    n_layers = 3, n_latent = 30, dispersion = 'gene-batch',\n",
    "                    # Leiden clustering\n",
    "                    leiden_clustering = None, col_cell_type = ['taa_l4'], \n",
    "                    fig_dir = f'{plots_path}/phenoAnalysis/trajectories/', fig_prefix = f'thyAgeing_dnInnateSplit_scvi_{object_version}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scvi_run['data'].obs['path_cellranger_arc'].dtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save adata and scvi model\n",
    "overwrite = True\n",
    "\n",
    "for c in scvi_run['data'].obs.columns:\n",
    "    if scvi_run['data'].obs[c].dtype == 'object':\n",
    "        scvi_run['data'].obs[c] = scvi_run['data'].obs[c].astype(str)\n",
    "    if scvi_run['data'].obs[c].dtype == 'category':\n",
    "        scvi_run['data'].obs[c] = scvi_run['data'].obs[c].astype(str)\n",
    "        \n",
    "anno_cols = [c for c in scvi_run['data'].obs.columns if '_pred_' in c or '_prob_' in c or 'taa' in c]\n",
    "if not os.path.exists(f'{data_path}/thyAgeing_dnInnateSplit_scvi_{object_version}.zarr') or overwrite:\n",
    "    scvi_run['data'].obs = scvi_run['data'].obs.drop(columns=anno_cols)\n",
    "    scvi_run['data'].write_h5ad(\n",
    "        f'{data_path}/objects/rna/thyAgeing_dnInnateSplit_scvi_{object_version}.zarr',\n",
    "        compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "        compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,\n",
    "    )\n",
    "    scvi_run['vae'].save(f'{model_path}/thyAgeing_dnInnateSplit_scvi_{object_version}', save_anndata=False, overwrite=overwrite)\n",
    "else:\n",
    "    print('File already exists')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run Palantir (diffusion mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import palantir \n",
    "use_rep_diffusion = 'X_umap'\n",
    "n_diffusion_comp = 30\n",
    "dm_res = palantir.utils.run_diffusion_maps(adata_sub, pca_key=use_rep_diffusion, n_components=n_diffusion_comp)#palantir.utils.run_diffusion_maps(pd.DataFrame(adata_sub.obsm[use_rep_diffusion]), n_components=n_diffusion_comp)\n",
    "plt.scatter(np.arange(n_diffusion_comp), dm_res['EigenValues'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate neighbor graph in multiscale diffusion space\n",
    "ms_data = palantir.utils.determine_multiscale_space(adata_sub,n_eigs=11)\n",
    "adata_sub.obsm[\"X_palantir\"]= ms_data.values\n",
    "sc.pp.neighbors(adata_sub,n_neighbors=20,use_rep=\"X_palantir\")\n",
    "\n",
    "with plt.rc_context({\"figure.figsize\": calc_figsize(width = 60, height = 35)}):\n",
    "    sc.pl.embedding(adata_sub, basis='X_palantir', color=[col_cell_type_fine, col_age_group], legend_loc='on data', return_fig = True)\n",
    "    plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_palantirEmbed_umap.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Impute gene expression values\n",
    "imputed_X = palantir.utils.run_magic_imputation(adata_sub, dm_res = dm_res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct_markers = {'ETP' : ['CD34'],\n",
    "              'T lin' : ['CD3D', 'CD3E', 'NOTCH1', 'TCF7', 'PTCRA', 'RAG1', 'RAG2', 'DNTT'],\n",
    "              'B lin' : ['VPREB1', 'IFITM3', 'IGLL1', 'CD79A', #'POU2AF1', 'PAX5' #-> not really expressed\n",
    "                         ],\n",
    "              'ILC' : ['ID2', 'ZBTB16', 'NFIL3', 'IL7R', 'KIT', 'KLRB1'],\n",
    "              'NK' : ['EOMES', 'NCAM1', 'PRF1', 'KLRD1'],\n",
    "              'pDC lin' : ['LILRA4', 'CLEC4C', 'RUNX2', 'TYROBP']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with plt.rc_context({\"figure.figsize\": calc_figsize(width = 60, height = 35)}):\n",
    "    sc.pl.DotPlot(adata_sub, \n",
    "                var_names=ct_markers, \n",
    "                groupby=col_cell_type_fine, \n",
    "                categories_order=col_cell_type_fine_levels,\n",
    "                #figsize = calc_figsize(width = 90, height = 0),\n",
    "                mean_only_expressed=True,\n",
    "                cmap = sns.blend_palette([thyAgeing_colors['blue'], thyAgeing_colors['purple'], thyAgeing_colors['magenta'], thyAgeing_colors['orange'], thyAgeing_colors['yellow']], as_cmap=True, n_colors=10),\n",
    "                ).add_totals().savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_linMarkers_dotplot.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub.obs.groupby([ col_cell_type_fine, 'age_group',]).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine].isin(['ILC', 'T_MAIT'])].obs.groupby(['age_group','donor'], observed=True).size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with plt.rc_context({\"figure.figsize\": calc_figsize(width = 80, height = 90)}):\n",
    "    sc.pl.embedding(\n",
    "        adata_sub,\n",
    "        basis=\"X_umap\",\n",
    "        layer=\"MAGIC_imputed_data\",\n",
    "        color=[g for sublist in ct_markers.values() for g in sublist],\n",
    "        cmap = sns.blend_palette([thyAgeing_colors['blue'], thyAgeing_colors['purple'], thyAgeing_colors['magenta'], thyAgeing_colors['orange'], thyAgeing_colors['yellow']], as_cmap=True, n_colors=10),\n",
    "        ncols = 5,\n",
    "        frameon=False,\n",
    "        return_fig=True,    \n",
    "        show = False\n",
    "    )\n",
    "    plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_lineageMarkers_imputed_umap.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import gaussian_kde\n",
    "\n",
    "def select_high_expr_cell(adata, gene_name : str) :\n",
    "    return adata[adata[:, gene_name].X.argmax()].obs_names[0]\n",
    "\n",
    "def calculate_centroid(adata, cell_population, embedding='X_umap'):\n",
    "    \"\"\"\n",
    "    Calculate the point of maximum density in x and y directions for a specific cell population in a given embedding \n",
    "    and return the barcode of the cell closest to this point.\n",
    "\n",
    "    Parameters:\n",
    "    adata (anndata.AnnData): The annotated data matrix.\n",
    "    cell_population (str): The cell population to calculate the point of maximum density for.\n",
    "    embedding (str): The embedding to use for calculating the point of maximum density. Default is 'X_umap'.\n",
    "\n",
    "    Returns:\n",
    "    str: The barcode of the cell closest to the point of maximum density.\n",
    "    \"\"\"\n",
    "\n",
    "    # Filter the data for the specific cell population\n",
    "    subset = adata[adata.obs[col_cell_type_fine] == cell_population]\n",
    "    \n",
    "    # Calculate the density\n",
    "    kde = gaussian_kde(subset.obsm[embedding].T)\n",
    "    density = kde(subset.obsm[embedding].T)\n",
    "    \n",
    "    # Find the cell with maximum density\n",
    "    max_density_idx = np.argmax(density)\n",
    "    \n",
    "    return subset.obs_names[max_density_idx]\n",
    "\n",
    "# Determine early cell\n",
    "start_cell = select_high_expr_cell(adata_sub[adata_sub.obs[col_cell_type_fine] == 'T_DN(early)'], 'CD34')\n",
    "#start_cell = calculate_centroid(adata_sub, 'T_DN(early)')\n",
    "\n",
    "# Determine terminal states\n",
    "terminal_cells = pd.Series(\n",
    "    [\"T_DN(P)\", \"B_dev_thy\", \"pDC\", 'ILC', 'NK_tr'],\n",
    "    index=[calculate_centroid(adata_sub, 'T_DN(P)'), \n",
    "           select_high_expr_cell(adata_sub[adata_sub.obs[col_cell_type_fine] == 'B_dev_thy'], 'IFITM3'), \n",
    "           select_high_expr_cell(adata_sub[adata_sub.obs[col_cell_type_fine] == 'pDC'], 'CLEC4C'),\n",
    "           select_high_expr_cell(adata_sub[adata_sub.obs[col_cell_type_fine] == 'ILC'], 'KIT'),\n",
    "           select_high_expr_cell(adata_sub[adata_sub.obs[col_cell_type_fine] == 'NK_tr'], 'EOMES')\n",
    "        #   adata_sub.obs_names[adata_sub.obsm['X_umap'][:,0].argmax()], \n",
    "        #    adata_sub.obs_names[adata_sub.obsm['X_umap'][:,1].argmin()], \n",
    "        #    adata_sub.obs_names[adata_sub.obsm['X_umap'][:,0].argmin()]],\n",
    "])\n",
    "\n",
    "all_cells = terminal_cells.copy()\n",
    "all_cells[start_cell] = 'T_DN(early)'\n",
    "palantir.plot.highlight_cells_on_umap(adata_sub, all_cells)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run palantir\n",
    "pr_res = palantir.core.run_palantir(\n",
    "    adata_sub, \n",
    "    early_cell=start_cell,\n",
    "    terminal_states=terminal_cells,\n",
    "    knn = 20,\n",
    "    num_waypoints=500, \n",
    "    use_early_cell_as_start=True,\n",
    "    seed = 20,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import palantir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = palantir.plot.plot_palantir_results(adata_sub, s=3, \n",
    "                                        cmap = sns.blend_palette([thyAgeing_colors['blue'], thyAgeing_colors['purple'], thyAgeing_colors['magenta'], thyAgeing_colors['orange'], thyAgeing_colors['yellow']], as_cmap=True, n_colors=10))\n",
    "p.set_size_inches(calc_figsize(width = 120, height = 60))\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_pseudotime_palantir_blue.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select branches\n",
    "masks = palantir.presults.select_branch_cells(adata_sub, q=.05, eps=.005)\n",
    "\n",
    "palantir.plot.plot_branch_selection(adata_sub)\n",
    "p.set_size_inches(calc_figsize(width = 100, height = 100))\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_branches_palantir.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot trajectories\n",
    "fig, axes = plt.subplots(1,3, figsize=calc_figsize(width = 80, height = 30))\n",
    "cmap = sns.blend_palette([thyAgeing_colors['teal'], thyAgeing_colors['yellow'], thyAgeing_colors['orange'], thyAgeing_colors['magenta']], as_cmap=True, n_colors=10)\n",
    "\n",
    "# Plot for B_dev_thy\n",
    "palantir.plot.plot_trajectory(\n",
    "    adata_sub,\n",
    "    \"B_dev_thy\",\n",
    "    cell_color=\"palantir_entropy\",\n",
    "    n_arrows=5,\n",
    "    color=\"black\",\n",
    "    scanpy_kwargs=dict(cmap=cmap),\n",
    "    arrowprops=dict(arrowstyle=\"-|>,head_length=.05,head_width=.05\"),\n",
    "    ax=axes[0]\n",
    ")\n",
    "axes[0].set_title(\"B_dev_thy\")\n",
    "\n",
    "# Plot for T_DN(P)\n",
    "palantir.plot.plot_trajectory(\n",
    "    adata_sub,\n",
    "    \"T_DN(P)\",\n",
    "    cell_color=\"palantir_entropy\",\n",
    "    n_arrows=5,\n",
    "    color=\"black\",\n",
    "    scanpy_kwargs=dict(cmap=cmap),\n",
    "    arrowprops=dict(arrowstyle=\"-|>,head_length=.05,head_width=.05\"),\n",
    "    ax=axes[1]\n",
    ")\n",
    "axes[1].set_title(\"T_DN(P)\")\n",
    "\n",
    "# Plot for pDC\n",
    "palantir.plot.plot_trajectory(\n",
    "    adata_sub,\n",
    "    \"pDC\",\n",
    "    cell_color=\"palantir_entropy\",\n",
    "    n_arrows=5,\n",
    "    color=\"black\",\n",
    "    scanpy_kwargs=dict(cmap=cmap),\n",
    "    arrowprops=dict(arrowstyle=\"-|>,head_length=.05,head_width=.05\"),\n",
    "    ax=axes[2]\n",
    ")\n",
    "axes[2].set_title(\"pDC\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_branches_palantir_umap.pdf', bbox_inches='tight', dpi = 300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inspect gene trends\n",
    "gene_trends = palantir.presults.compute_gene_trends(\n",
    "    adata_sub,\n",
    "    expression_key=\"MAGIC_imputed_data\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p = palantir.plot.plot_gene_trends(adata_sub, [g for sublist in ct_markers.values() for g in sublist],)\n",
    "p.set_size_inches(calc_figsize(width = 60, height = 247))\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_lineageMarkers_palantir_lines.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as mpl\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "norm = mpl.colors.Normalize(vmin=0, vmax=2)\n",
    "goi = [g for sublist in ct_markers.values() for g in sublist]\n",
    "goi.remove('RAG1')\n",
    "goi.remove('RAG2')\n",
    "goi.remove('CD3E')\n",
    "p = palantir.plot.plot_gene_trend_heatmaps(\n",
    "    adata_sub, \n",
    "    genes=goi,\n",
    "    scaling='none',\n",
    "    gene_trend_key='gene_trends',\n",
    "    cmap=sns.blend_palette(\n",
    "        [thyAgeing_colors['blue'], thyAgeing_colors['purple'], thyAgeing_colors['magenta'], thyAgeing_colors['orange'], thyAgeing_colors['yellow']],\n",
    "        as_cmap=True, n_colors=10\n",
    "    ),\n",
    "    basefigsize=calc_figsize(width=30, height=50),\n",
    "    branch_names=[\"B_dev_thy\", \"T_DN(P)\", \"pDC\"],\n",
    "    norm=norm,\n",
    "    cbkwargs={'label': 'log2FC', 'ticks': np.linspace(0, 2, 5), 'aspect': 30, 'pad': 0.05}\n",
    ")\n",
    "\n",
    "# Adjust colorbar height to match the plot\n",
    "fig = p if isinstance(p, mpl.figure.Figure) else plt.gcf()\n",
    "for ax in fig.axes:\n",
    "    # Find colorbar axes\n",
    "    if hasattr(ax, 'get_ylabel') and ax.get_ylabel() == 'log2FC':\n",
    "        pos = ax.get_position()\n",
    "        ax.set_position([pos.x0, pos.y0, pos.width, fig.axes[0].get_position().height])\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_lineageMarkers_palantir_heatmap.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add vdj data\n",
    "meta_tcr = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_v9_2025-03-28_tcrgd_v6.csv', index_col = 0)\n",
    "adata_sub.obs = adata_sub.obs.join(meta_tcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add vdj data\n",
    "meta_tcr = pd.read_csv(f'{data_path}/objects/rna/thyAgeing_tSplit_scvi_v9_2025-03-28_tcrgd_v6.csv', index_col = 0)\n",
    "adata.obs = adata.obs.join(meta_tcr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub.obs.drop(columns = meta_tcr.columns, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_tcr.loc[meta_tcr['productive_VDJ'] != 'No_contig']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'ILC'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_tcr.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'ILC'].obs[['d_call_VDJ_main', 'j_call_VDJ_main']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'T_MAIT'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'NK_tr'].obs[['d_call_VDJ_main', 'j_call_VDJ_main']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'pDC'].obs[['d_call_VDJ_main', 'j_call_VDJ_main']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'NK_tr'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'pDC'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'B_dev_thy'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub[adata_sub.obs[col_cell_type_fine] == 'T_DN(P)'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata[adata.obs[col_cell_type_fine] == 'T_DN(Q)'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata[adata.obs[col_cell_type_fine] == 'T_DP(P)'].obs[['productive_VDJ']].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata[adata.obs[col_cell_type_fine] == 'T_DP(Q)'].obs[['productive_VJ']].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression of gene clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct = 'T_DN(P)'\n",
    "\n",
    "# Determine gene clusters\n",
    "highly_variable_genes = list(set(adata_sub.var[adata_sub.var['highly_variable']].index.tolist() + goi))\n",
    "communities = palantir.presults.cluster_gene_trends(adata_sub, ct, highly_variable_genes, n_neighbors=30)\n",
    "\n",
    "palantir.plot.plot_gene_trend_clusters(adata_sub, ct)\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_geneClusters_{ct}_palantir_lineplot.pdf', bbox_inches='tight', dpi = 300)\n",
    "\n",
    "pd.DataFrame(communities, columns=['cluster']).reset_index(names = 'gene').sort_values(by = 'cluster').to_excel(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_geneClusters_{ct}_palantir.xlsx'.format(ct), index = False, sheet_name=ct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "community_dict = {f'cluster_{category}': communities[communities == category].index.tolist() for category in communities.cat.categories}\n",
    "import pprint\n",
    "pprint.pprint(community_dict, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fate probabilities in ETP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fate probabilities\n",
    "df = adata.obs[['donor', 'sample', 'age_group', col_cell_type_fine]].copy()\n",
    "df = df.join(adata_sub.obsm['palantir_fate_probabilities'])\n",
    "df.dropna(subset = 'pDC', inplace = True)\n",
    "df = df.melt(id_vars = ['donor', 'sample', 'age_group', col_cell_type_fine], var_name = 'cell_type', value_name = 'fate_prob')\n",
    "df = df.groupby(['donor', 'sample', 'age_group', col_cell_type_fine, 'cell_type'], observed = True)['fate_prob'].mean().reset_index()\n",
    "df = df.groupby(['donor', 'age_group', col_cell_type_fine, 'cell_type'], observed = True)['fate_prob'].mean().reset_index()\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from plotting.utils import plot_grouped_boxplot,get_tint_palette\n",
    "df['cell_type'] = df['cell_type'].str.replace('T_', '').str.replace('_thy', '')\n",
    "plot_grouped_boxplot(data = df.loc[df[col_cell_type_fine] == 'T_DN(early)'], x = 'cell_type', y = 'fate_prob', hue = col_age_group, order = ['DN(P)', 'B_dev', 'pDC'], hue_order = ['infant', 'paed', 'adult', 'aged'],\n",
    "                     palette =  get_tint_palette(thyAgeing_colors['magenta']),\n",
    "                     x_label = 'Cell population', y_label = 'Fate probability', legend_title = 'Age group', add_stats = True, format_percent = False, figsize = calc_figsize(width = 40, height = 45),\n",
    "                     save_stats = f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_fateProbs_palantir_',\n",
    "                     legend_kwargs = {'bbox_to_anchor':(1.05, 1), 'loc':'upper left'}, ylim = (0,1),\n",
    "                     )\n",
    "#plt.xticks(rotation=45, ha='right')\n",
    "plt.yticks([0, 0.5, 1])\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_fateProbs_palantir_boxplot.pdf', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in adata_sub.obs.columns:\n",
    "        if pd.api.types.is_object_dtype(adata_sub.obs[c]):\n",
    "                adata_sub.obs[c] = adata_sub.obs[c].astype(str)\n",
    "                \n",
    "adata_sub.write_h5ad(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_palantir.zarr',\n",
    "                compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "                compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata_sub = ad.read_h5ad(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_dnInnateSplit_palantir_.zarr')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## scFates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Remove cell cycle genes\n",
    "# cell_cycle_genes = [x.strip() for x in open('/nfs/team205/vk8/processed_data/regev_lab_cell_cycle_genes.txt')]\n",
    "# adata_sub = adata_sub[:,~adata_sub.var.index.isin(cell_cycle_genes)]\n",
    "\n",
    "# # Log-normalize\n",
    "# adata_sub.layers['counts'] = adata_sub.X.copy()\n",
    "# sc.pp.normalize_total(adata_sub, target_sum=1e4)\n",
    "# sc.pp.log1p(adata_sub)\n",
    "\n",
    "# # Select HVG and run PCA\n",
    "# sc.pp.highly_variable_genes(adata_sub, n_top_genes=500, flavor='seurat', n_bins=20, batch_key='study')\n",
    "# sc.pp.pca(adata_sub, n_comps=20, use_highly_variable=True, svd_solver='arpack')\n",
    "\n",
    "# sc.pl.pca(adata_sub, color = [col_cell_type_fine, col_age_group], ncols=2, wspace=0.5)\n",
    "\n",
    "# # Perform UMAP\n",
    "# adata_sub.obsm['X_umap_scVI'] = adata_sub.obsm['X_umap'].copy()\n",
    "# sc.pp.neighbors(adata_sub, n_pcs=9, n_neighbors=20, random_state=42)\n",
    "# sc.tl.umap(adata_sub)\n",
    "\n",
    "# sc.pl.umap(adata_sub, color=[col_cell_type_fine, col_age_group], ncols=2, wspace=0.5, return_fig=True)\n",
    "# plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_lateSplit_umap.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import palantir \n",
    "use_rep_diffusion = 'X_scVI'\n",
    "n_diffusion_comp = 30\n",
    "dm_res = palantir.utils.run_diffusion_maps(adata_sub, pca_key=use_rep_diffusion, n_components=n_diffusion_comp)#palantir.utils.run_diffusion_maps(pd.DataFrame(adata_sub.obsm[use_rep_diffusion]), n_components=n_diffusion_comp)\n",
    "plt.scatter(np.arange(n_diffusion_comp), dm_res['EigenValues'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate neighbor draph in multiscale diffusion space\n",
    "ms_data = palantir.utils.determine_multiscale_space(adata_sub,n_eigs=10)\n",
    "adata_sub.obsm[\"X_palantir\"]= ms_data.values\n",
    "sc.pp.neighbors(adata_sub,n_neighbors=20,use_rep=\"X_palantir\")\n",
    "\n",
    "sc.pl.embedding(adata_sub, basis='X_palantir', color=[col_cell_type_fine, col_age_group], legend_loc='on data', return_fig = True)\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_etpSplit_palantirEmbed_umap.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# draw ForceAtlas2 embedding using 2 first PCs as initial positions\n",
    "adata_sub.obsm[\"X_palantir2d\"]=adata_sub.obsm[\"X_palantir\"][:,:2]\n",
    "sc.tl.draw_graph(adata_sub,init_pos='X_palantir2d', n_jobs = 64)\n",
    "\n",
    "sc.pl.draw_graph(adata_sub,color=col_cell_type_fine,legend_loc='on data')\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_etpSplit_palantirEmbed_forceAtlas.pdf', bbox_inches='tight', dpi = 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in adata_sub.obs.columns:\n",
    "        if pd.api.types.is_object_dtype(adata_sub.obs[c]) or isinstance(adata_sub.obs[c].dtype, pd.CategoricalDtype):\n",
    "                adata_sub.obs[c] = adata_sub.obs[c].astype(str)\n",
    "                \n",
    "adata_sub.write_h5ad(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_etpSplit_palantir_.zarr',\n",
    "                compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "                compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,\n",
    "        )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub = ad.read_h5ad(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_etpSplit_palantir_.zarr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scf.tl.tree(adata_sub,method=\"ppt\",Nodes=80,use_rep=\"X_palantir\",\n",
    "            device=\"cpu\",seed=1,ppt_lambda=100,ppt_sigma=0.025,ppt_nsteps=200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scf.pl.graph(adata_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scf.tl.root(adata_sub,21)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scf.tl.pseudotime(adata_sub,n_jobs=20,n_map=100,seed=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scf.pl.trajectory(adata_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.draw_graph(adata_sub,color=[\"seg\",\"milestones\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scf.tl.rename_milestones(adata_sub,['BifA', 'BifB', 'T_Treg', 'T_αβT(entry)', 'BifC', 'T_CD8_naive','T_Treg(agonist)', 'T_CD4_naive'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "scf.pl.milestones(adata_sub, annotate=True, show = False)\n",
    "plt.savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_milestones.pdf', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for c in adata_sub.obs.columns:\n",
    "#         if pd.api.types.is_object_dtype(adata_sub.obs[c]) or isinstance(adata_sub.obs[c].dtype, pd.CategoricalDtype):\n",
    "#                 adata_sub.obs[c] = adata_sub.obs[c].astype(str)\n",
    "        \n",
    "# adata_sub.write_h5ad(f'{data_path}/objects/rna/thyAgeing_lateSplit_palantir.zarr',\n",
    "#                      compression=hdf5plugin.FILTERS[\"zstd\"],\n",
    "#                      compression_opts=hdf5plugin.Zstd(clevel=5).filter_options,)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load adata with trajectory\n",
    "adata_sub = ad.read_h5ad(f'{data_path}/objects/rna/thyAgeing_lateSplit_palantir.zarr')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Log-normalize\n",
    "adata_sub.layers['counts'] = adata_sub.X.copy()\n",
    "sc.pp.normalize_total(adata_sub, target_sum=1e4)\n",
    "sc.pp.log1p(adata_sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_expressed_genes(adata: ad.AnnData, groupby : str, group : str, min_frac : float = 0.1) -> pd.Series:\n",
    "    \n",
    "    n_obs_group = (adata.obs[groupby] == group).sum()\n",
    "    ecf = np.array(adata[adata.obs[groupby] == group].X.astype(bool).sum(axis=0) / n_obs_group).flatten()\n",
    "    \n",
    "    return list(adata.var_names[ecf > min_frac])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub.obs['t']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### abT(entry) -> mature T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub_seg = adata_sub[(adata_sub.obs['seg'] == '5') & (adata_sub.obs[col_age_group].isin(['adult', 'infant']))].copy()\n",
    "adata_sub_seg.obs[col_age_group] = pd.Categorical(adata_sub_seg.obs[col_age_group], categories = ['infant', 'adult'], ordered=True)\n",
    "features_to_test = list(set(get_expressed_genes(adata_sub_seg, col_age_group, 'infant') + get_expressed_genes(adata_sub_seg, col_age_group, 'adult')))\n",
    "\n",
    "print(f'Number of features to test: {len(features_to_test)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amplitute test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=False, fdr_cut=0.05, n_jobs=4, n_map=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seg5_degs = adata_sub_seg.var.copy()\n",
    "seg5_degs['log2FC'] = - seg5_degs['infant->adult_lfc']\n",
    "\n",
    "seg5_degs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seg5_degs[seg5_degs['cov_fdr'] < .05].sort_values('log2FC').head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seg5_degs[seg5_degs['cov_fdr'] < .05].sort_values('log2FC', ascending=False).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seg5_degs.to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_seg5_amplitudeTest.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Trend test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=True, fdr_cut=0.05, n_jobs=4, n_map=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seg5_trend_degs = adata_sub_seg.var.copy()\n",
    "seg5_trend_degs['log2FC'] = - seg5_trend_degs['infant->adult_lfc']\n",
    "\n",
    "seg5_trend_degs.sort_values('log2FC').head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seg5_trend_degs.sort_values('log2FC', ascending=False).head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seg5_trend_degs.to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_seg5_trendTest.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Mature T cells"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub_seg = adata_sub[(adata_sub.obs['milestones'] == 'T_CD4_naive') & (adata_sub.obs[col_age_group].isin(['adult', 'infant']))].copy()\n",
    "adata_sub_seg.obs[col_age_group] = pd.Categorical(adata_sub_seg.obs[col_age_group], categories = ['infant', 'adult'], ordered=True)\n",
    "features_to_test = list(set(get_expressed_genes(adata_sub_seg, col_age_group, 'infant') + get_expressed_genes(adata_sub_seg, col_age_group, 'adult')))\n",
    "\n",
    "print(f'Number of features to test: {len(features_to_test)}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Amplitute test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=False, fdr_cut=0.05, n_jobs=4, n_map=1)\n",
    "\n",
    "degs = adata_sub_seg.var.copy()\n",
    "degs = degs.loc[~degs['cov_signi'].isna()]\n",
    "degs['log2FC'] = - degs['infant->adult_lfc']\n",
    "\n",
    "degs.sort_values('infant->adult_lfc').to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_CD4_naive_amplitudeTest.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Trend test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=True, fdr_cut=0.05, n_jobs=4, n_map=1)\n",
    "\n",
    "degs = adata_sub_seg.var.copy()\n",
    "degs = degs.loc[~degs['cov_signi'].isna()]\n",
    "degs['log2FC'] = - degs['infant->adult_lfc']\n",
    "\n",
    "degs.sort_values('infant->adult_lfc').to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_CD4_naive_trendTest.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub_seg = adata_sub[(adata_sub.obs['milestones'] == 'T_CD8_naive') & (adata_sub.obs[col_age_group].isin(['adult', 'infant']))].copy()\n",
    "adata_sub_seg.obs[col_age_group] = pd.Categorical(adata_sub_seg.obs[col_age_group], categories = ['infant', 'adult'], ordered=True)\n",
    "features_to_test = list(set(get_expressed_genes(adata_sub_seg, col_age_group, 'infant') + get_expressed_genes(adata_sub_seg, col_age_group, 'adult')))\n",
    "\n",
    "print(f'Number of features to test: {len(features_to_test)}')\n",
    "\n",
    "# Amplitute test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=False, fdr_cut=0.05, n_jobs=4, n_map=1)\n",
    "\n",
    "degs = adata_sub_seg.var.copy()\n",
    "degs = degs.loc[~degs['cov_signi'].isna()]\n",
    "degs['log2FC'] = - degs['infant->adult_lfc']\n",
    "\n",
    "degs.sort_values('infant->adult_lfc').to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_CD8_naive_amplitudeTest.csv')\n",
    "\n",
    "# Trend test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=True, fdr_cut=0.05, n_jobs=4, n_map=1)\n",
    "\n",
    "degs = adata_sub_seg.var.copy()\n",
    "degs = degs.loc[~degs['cov_signi'].isna()]\n",
    "degs['log2FC'] = - degs['infant->adult_lfc']\n",
    "\n",
    "degs.sort_values('infant->adult_lfc').to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_CD8_naive_trendTest.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sub_seg = adata_sub[(adata_sub.obs['milestones'] == 'T_Treg') & (adata_sub.obs[col_age_group].isin(['adult', 'infant']))].copy()\n",
    "adata_sub_seg.obs[col_age_group] = pd.Categorical(adata_sub_seg.obs[col_age_group], categories = ['infant', 'adult'], ordered=True)\n",
    "features_to_test = list(set(get_expressed_genes(adata_sub_seg, col_age_group, 'infant') + get_expressed_genes(adata_sub_seg, col_age_group, 'adult')))\n",
    "\n",
    "print(f'Number of features to test: {len(features_to_test)}')\n",
    "\n",
    "# Amplitute test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=False, fdr_cut=0.05, n_jobs=4, n_map=1)\n",
    "\n",
    "degs = adata_sub_seg.var.copy()\n",
    "degs = degs.loc[~degs['cov_signi'].isna()]\n",
    "degs['log2FC'] = - degs['infant->adult_lfc']\n",
    "\n",
    "degs.sort_values('infant->adult_lfc').to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_Treg_amplitudeTest.csv')\n",
    "\n",
    "# Trend test\n",
    "scf.tl.test_covariate(adata_sub_seg, features = features_to_test, group_key = col_age_group, trend_test=True, fdr_cut=0.05, n_jobs=4, n_map=1)\n",
    "\n",
    "degs = adata_sub_seg.var.copy()\n",
    "degs = degs.loc[~degs['cov_signi'].isna()]\n",
    "degs['log2FC'] = - degs['infant->adult_lfc']\n",
    "\n",
    "degs.sort_values('infant->adult_lfc').to_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_Treg_trendTest.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load all DEG\n",
    "cd8_deg = pd.read_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_CD8_naive_trendTest.csv')\n",
    "cd4_deg = pd.read_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_CD4_naive_trendTest.csv')\n",
    "treg_deg = pd.read_csv(f'{data_path}/analyses/phenoAnalysis/trajectories/thyAgeing_lateSplit_scFates_T_Treg_trendTest.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersect(*d):\n",
    "    sets = iter(map(set, d))\n",
    "    result = sets.next()\n",
    "    for s in sets:\n",
    "        result = result.intersection(s)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from supervenn import supervenn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "up_sets = [set(cd8_deg.loc[cd8_deg['log2FC'] > 1]['gene_name'].tolist()), set(cd4_deg.loc[cd4_deg['log2FC'] > 1]['gene_name'].tolist()), set(treg_deg.loc[treg_deg['log2FC'] > 1]['gene_name'].tolist())]\n",
    "up_p = supervenn(up_sets, set_annotations=['CD8', 'CD4', 'Treg'])\n",
    "\n",
    "up_p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pprint\n",
    "pprint.pprint(up_p.chunks, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Hypoxia: ANKRD37\n",
    "- CXCR4: CXCR4 downregulation correlates with TCR-mediated signalling in DP thymocytes, resulting in overall donwregulation in mature CD4/CD8 T cells ([Lucas, 2017](https://www.nature.com/articles/s41598-017-05182-7))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "down_sets = [set(cd8_deg.loc[cd8_deg['log2FC'] < -1]['gene_name'].tolist()), set(cd4_deg.loc[cd4_deg['log2FC'] < -1]['gene_name'].tolist()), set(treg_deg.loc[treg_deg['log2FC'] < -1]['gene_name'].tolist())]\n",
    "down_p = supervenn(down_sets, set_annotations=['CD8', 'CD4', 'Treg'])\n",
    "\n",
    "down_p"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pprint.pprint(down_p.chunks, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- CD38 downregulation: CD38 is marker of RTEs ([Bohacova, 2024](https://www.cell.com/immunity/fulltext/S1074-7613(24)00418-7)) -> paper also shows age-related decline in CD38 in naive T cells in blood\n",
    "- SIRPBlow CD8 T cells have lower activation threshold and are implicated in autoimmunity ([Sinha, 2018](https://www.nature.com/articles/s41598-018-33901-1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "sc.pl.DotPlot(adata_sub[adata_sub.obs[col_cell_type_fine].isin(['T_CD8_naive', 'T_CD4_naive', 'T_Treg'])],\n",
    "              var_names = ['S1PR1', 'CD38', 'CXCR4', 'SIRPG'], groupby = [col_cell_type_fine, col_age_group], \n",
    "              categories_order= [f'{c1}_{c2}' for c1,c2 in list(itertools.product(['T_CD8_naive', 'T_CD4_naive', 'T_Treg'], col_age_group_levels))],\n",
    "              mean_only_expressed=True, \n",
    "              cmap = 'magma').add_totals().savefig(f'{plots_path}/phenoAnalysis/trajectories/thyAgeing_lateSplit_exitDegs_dotplot.pdf', dpi = 300, bbox_inches = 'tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "session_info.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
